<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>rCASC vignette • rCASC</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="rCASC vignette">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rCASC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/rCASC_vignette.html">rCASC vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>rCASC vignette</h1>
                        <h4 class="author">Luca Alessandrì, Francesca Cordero, Marco Beccuti, Maddalena Arigoni, Martina Olivero, Greta Romano, Sergio Rabellino, Gennaro De Libero, Luigia Pace, and Raffaele A Calogero</h4>
            
            <h4 class="date">26/12/2018</h4>
      
      
      <div class="hidden name"><code>rCASC_vignette.Rmd</code></div>

    </div>

    
    


<div id="section-1-rcasc-a-single-cell-analysis-workflow-designed-to-provide-data-reproducibility" class="section level2">
<h2 class="hasAnchor">
<a href="#section-1-rcasc-a-single-cell-analysis-workflow-designed-to-provide-data-reproducibility" class="anchor"></a><strong>Section 1</strong> rCASC: a single cell analysis workflow designed to provide data reproducibility</h2>
<p>Since the end of the 90’s omics high-throughput technologies have generated an enormous amount of data, reaching today an exponential growth phase. Analysis of omics big data is a revolutionary means of understanding the molecular basis of disease regulation and susceptibility, and this resource is accessible to the biological/medical community via bioinformatics frameworks. However, because of the fast evolution of computation tools and omics methods, the <a href="https://en.wikipedia.org/wiki/Replication_crisis"><em>reproducibility crisis</em></a> is becoming a very important issue [<a href="https://www.nature.com/collections/prbfkwmwvz"><em>Nature, 2018</em></a>] and there is a mandatory need to guarantee robust and reliable results to the research community [<a href="http://www.global-engage.com/life-science/reproducibility-computational-biology/"><em>Global Engage Blog</em></a>].</p>
<p>Our group is deeply involved in developing workflows that guarantee both <strong>functional</strong> (i.e. the information about data and the utilized tools are saved in terms of meta-data) and <strong>computation</strong> reproducibility (i.e. the real image of the computation environment used to generate the data is stored). For this reason we are managing a bioinformatics community called <a href="http://www.reproducible-bioinformatics.org/"><em>reproducible-bioinformatics.org</em></a> [<a href="https://rdcu.be/9gMq"><em>Kulkarni et al.</em></a>] designed to provide to the biological community a reproducible bioinformatics ecosystem [<a href="https://academic.oup.com/bioinformatics/article/34/5/871/4562334"><em>Beccuti et al.</em></a>].</p>
<p>rCASC, reproducible Cluster Analysis of Single Cells, is part of the <a href="http://www.reproducible-bioinformatics.org/"><em>reproducible-bioinformatics.org</em></a> project and provides single cell analysis functionalities within the reproducible rules described by Sandve et al. [<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285"><em>PLoS Comp Biol. 2013</em></a>]. rCASC is designed to provide a workflow for cell-subpopulation discovery:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/rcasc_workflow.jpg" alt="rCASC workflow" width="70%"><p class="caption">
rCASC workflow
</p>
</div>
<p>The workflow allows the direct analysis of fastq files generated with <a href="https://www.10xgenomics.com/"><em>10X Genomics platform</em></a>, <a href="https://1cell-bio.com/"><em>InDrop technology</em></a> or a count matrix having as column-names cells identifier and as row names ENSEMBL gene annotation. In the following paragraphs the functionalities of rCASC workflow are described.</p>

<div id="section-1-1-minimal-hardware-requirements-to-run-rcasc" class="section level3">
<h3 class="hasAnchor">
<a href="#section-1-1-minimal-hardware-requirements-to-run-rcasc" class="anchor"></a><strong>Section 1.1</strong> Minimal hardware requirements to run rCASC</h3>
<p>The RAM and CPU requirements are dependent on the data set under analysis, e.g. 500-600 cells can be effectively analysed using the hardware described by Beccuti [<a href="https://academic.oup.com/bioinformatics/article/34/5/871/4562334"><em>Bioinformatics 2018</em></a>]:</p>
<ul>
<li><p>32 Gb RAM</p></li>
<li><p>2.6 GHz Core i7 6700HQ with 8 threads.</p></li>
<li><p>500 GB SSD</p></li>
</ul>
<p>The analysis time can be significantly improved increasing the number of cores. Implementation of the workflow for computers farm, using <a href="https://docs.docker.com/engine/swarm/"><em>swarm</em></a>, is under implementation.</p>
</div>
<div id="section-1-2-installation" class="section level3">
<h3 class="hasAnchor">
<a href="#section-1-2-installation" class="anchor"></a><strong>Section 1.2</strong> Installation</h3>
<p>The minimal requirements for installation are:</p>
<ul>
<li><p>A workstation/server running 64 bits Linux.</p></li>
<li>
<p>Docker daemon installed on the machine, for more info see this document:</p>
<ul>
<li>
<a href="https://docs.docker.com/engine/installation/"><em>https://docs.docker.com/engine/installation/</em></a>.</li>
</ul>
</li>
<li><p>A scratch folder, e.g. /data/scratch, possibly on a fast I/O SSD disk, writable by everybody:</p></li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chmod <span class="dv">777</span> <span class="op">/</span>data<span class="op">/</span>scratch</code></pre></div>
<p></p>
<p>The functions in rCASC package require that user belongs to a <em>group</em> with the rights to execute docker. See the following document for more info: <a href="https://docs.docker.com/install/linux/linux-postinstall/"><em>https://docs.docker.com/install/linux/linux-postinstall/</em></a></p>
<p>The following commands allow the rCASC installation in <a href="https://cran.r-project.org/"><em>R</em></a> environment:</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"devtools"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(devtools)
<span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"kendomaniac/rCASC"</span>)

<span class="co"># downloading the required docker containers</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/downloadContainers.html">downloadContainers</a></span>()</code></pre></div>
<p></p>
</div>
</div>
<div id="section-2-counts-generation" class="section level2">
<h2 class="hasAnchor">
<a href="#section-2-counts-generation" class="anchor"></a><strong>Section 2</strong> Counts generation</h2>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui1.png" alt="GUI: Counts generation Menu" width="50%"><p class="caption">
GUI: Counts generation Menu
</p>
</div>
<p>This session refers to the generation of a counts table starting from fastq files generated by inDrop and 10XGenomics platforms.</p>

<div id="section-2-3-sra-download-new" class="section level3">
<h3 class="hasAnchor">
<a href="#section-2-3-sra-download-new" class="anchor"></a><strong>Section 2.3</strong> SRA download (New)</h3>
<p>The function <strong>sraDownload</strong> allows to download for NIH SRA repository fastq files.</p>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help):
<ul>
<li>
<em>group</em>, a character string. Two options: sudo or docker, depending to which group the user belongs</li>
<li>
<em>data.folder</em>, a character string indicating the working folder where output folder will be written (The output folder will contain the downloaded fastq file/s)</li>
<li>
<em>scratch.folder</em>, a character string indicating the temporary folder for data preprocessing</li>
<li>
<em>threads</em>, a integer indicating the number of threads to be used from fasterq-dumper</li>
</ul>
</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="co">#Downloading fastq from NCBI SRA</span>
<span class="kw"><a href="../reference/sraDownload.html">sraDownload</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">sra.name=</span><span class="st">"SRR7762358"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>, <span class="dt">threads=</span><span class="dv">8</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv ./SRR7762358/SRR7762358.fastq.gz ./SRR7762358/SRR7762358_S1_L001_R1_001.fastq.gz"</span>)</code></pre></div>
<p></p>
</div>
<div id="section-2-2-indrop-seq" class="section level3">
<h3 class="hasAnchor">
<a href="#section-2-2-indrop-seq" class="anchor"></a><strong>Section 2.2</strong> inDrop seq</h3>
<p>inDrop single-cell sequencing approach was originally published by Klein [<a href="https://www.ncbi.nlm.nih.gov/pubmed/26000487"><em>Cell 2015</em></a>]. Then, the authors published the detailed protocol in [<a href="https://www.ncbi.nlm.nih.gov/pubmed/27929523"><em>Zilionis et al. Nature Protocols 2017</em></a>], which has different primer design comparing to the original paper:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/indrop_v2.jpeg" alt="inDrop library structure" width="100%"><p class="caption">
inDrop library structure
</p>
</div>
<p>The analysis shown below is based on the protocol in Zilionis [<a href="https://www.ncbi.nlm.nih.gov/pubmed/27929523"><em>Nature Protocols 2017</em></a>], which is the version 2 of the inDrop technology, actually distributed by <a href="https://1cell-bio.com/product/indrop-system/"><em>1CellBio</em></a>:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/indropv2.jpg" alt="inDrop v2" width="100%"><p class="caption">
inDrop v2
</p>
</div>
<div id="section-2-2-1-indrop-data-analysis" class="section level4">
<h4 class="hasAnchor">
<a href="#section-2-2-1-indrop-data-analysis" class="anchor"></a><strong>Section 2.2.1</strong> inDrop data analysis</h4>
<p>There are two main functions, <strong>indropIndex</strong> and <strong>indropCounts</strong>, allowing the generation of a counts table starting from fastq files:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui2.png" alt="GUI: inDrop related panels: A) inDrop Index generation, B) inDrop counts table generation" width="50%"><p class="caption">
GUI: inDrop related panels: A) inDrop Index generation, B) inDrop counts table generation
</p>
</div>
<div id="section-2-2-1-1-indropindex-creates-a-reference-genome-for-indrop-v2" class="section level5">
<h5 class="hasAnchor">
<a href="#section-2-2-1-1-indropindex-creates-a-reference-genome-for-indrop-v2" class="anchor"></a><strong>Section 2.2.1.1</strong> indropIndex: Creates a reference genome for inDrop V2</h5>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above, panel A):
<ul>
<li><p><em>index.folder</em>: the folder where the reference genome will be created</p></li>
<li><p><em>ensembl.urlgenome</em>: the link to the ENSEMBL unmasked genome sequence of interest.</p></li>
<li><p><em>ensembl.urlgtf</em>: the link to the ENSEMBL GTF of the genome of interest.</p></li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="co">#running indropCounts index build</span>
<span class="kw"><a href="../reference/indropIndex.html">indropIndex</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">index.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),
 <span class="dt">ensembl.urlgenome=</span><span class="st">"ftp://ftp.ensembl.org/pub/release-87/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.toplevel.fa.gz"</span>,
 <span class="dt">ensembl.urlgtf=</span><span class="st">"ftp://ftp.ensembl.org/pub/release-87/gtf/mus_musculus/Mus_musculus.GRCm38.87.gtf.gz"</span>)</code></pre></div>
<p></p>
</div>
<div id="section-2-2-1-2-indropcounts-converts-fastq-in-umi-counts-using-indrop-workflow" class="section level5">
<h5 class="hasAnchor">
<a href="#section-2-2-1-2-indropcounts-converts-fastq-in-umi-counts-using-indrop-workflow" class="anchor"></a><strong>Section 2.2.1.2</strong> indropCounts: Converts fastq in UMI counts using <a href="https://github.com/indrops"><em>inDrop workflow</em></a>
</h5>
<ul>
<li>
<p><em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above, panel B):</p>
<ul>
<li>
<em>scratch.folder</em>, a character string indicating the path of the scratch folder</li>
<li>
<em>fastq.folder (GUI FastQ folder)</em>, a character string indicating the folder where input data are located and where output will be written</li>
<li>
<em>index.folder (GUI Index Folder)</em>, a character string indicating the folder where transcriptome index was created with indropIndex.</li>
<li>
<em>split.affixes</em>, the string separating SAMPLENAME from the Rz_001.fastq.gz, e.g. S0_L001.</li>
<li>
<em>bowtie.index.prefix</em>, the prefix name of the bowtie index. If genome was generated with indropIndex function the bowtie index is genome (default).</li>
<li>
<em>M</em>, integer. Ignore reads with more than M alignments, after filtering on distance from transcript end, suggested value 10</li>
<li>
<em>U</em>, integer. Ignore counts from UMI that should be split among more than U genes, suggested value 2</li>
<li>
<em>D</em>, integer. Maximal distance from transcript end, suggested valur 400.</li>
<li>
<em>low.complexity.mask</em>, boulean, masking low complexity regions</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget 130.192.119.59/public/testMm_S0_L001_R1_001.fastq.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget 130.192.119.59/public/testMm_S0_L001_R2_001.fastq.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/indropCounts.html">indropCounts</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>, <span class="dt">fastq.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),
        <span class="dt">index.folder=</span><span class="st">"/data/genomes/indropMm10"</span>, <span class="dt">sample.name=</span><span class="st">"testMm"</span>, <span class="dt">split.affixes=</span><span class="st">"S0_L001"</span>,
        <span class="dt">bowtie.index.prefix=</span><span class="st">"genome"</span>, <span class="dt">M=</span><span class="dv">10</span>, <span class="dt">U=</span><span class="dv">2</span>, <span class="dt">D=</span><span class="dv">400</span>, <span class="dt">low.complexity.mask=</span><span class="st">"False"</span>)</code></pre></div>
<p></p>
</div>
</div>
</div>
<div id="section-2-3-10xgenomics" class="section level3">
<h3 class="hasAnchor">
<a href="#section-2-3-10xgenomics" class="anchor"></a><strong>Section 2.3</strong> 10XGenomics</h3>
<p>The rCASC function, <strong>cellrangerCount</strong> allows the generation of a counts table starting from fastq files. This function implements <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview"><em>Cellranger</em></a>, the 10xGenomics tool allowing the conversion of the fastqs, generated with 10XGenomics platform, into a count matrix:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui3.png" alt="GUI: 10XGenomics counts table generation" width="50%"><p class="caption">
GUI: 10XGenomics counts table generation
</p>
</div>
<p>Genome indexes are retrieved from <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest">10Xgenomics repository</a>.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"/data/genomes/cellranger_hg38"</span>)
<span class="co">#getting the hg38 human genome cellranger index </span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget  http://cf.10xgenomics.com/supp/cell-exp/refdata-cellranger-GRCh38-3.0.0.tar.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"/data/genomes/cellranger_hg19"</span>)
<span class="co">#getting the hg19 human genome cellranger index</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget  http://cf.10xgenomics.com/supp/cell-exp/refdata-cellranger-hg19-3.0.0.tar.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"/data/genomes/cellranger_mm10"</span>)
<span class="co">#getting the mm10 mouse genome cellranger index</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://cf.10xgenomics.com/supp/cell-exp/refdata-cellranger-mm10-3.0.0.tar.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"/data/genomes/cellranger_hg19mm10"</span>)
<span class="co">#getting the human and mouse cellranger index</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://cf.10xgenomics.com/supp/cell-exp/refdata-cellranger-hg19-and-mm10-2.1.0.tar.gz"</span>)</code></pre></div>
<p></p>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>fastq (GUI FastQ folder)</em>, the path to the folder, where 10XGenomics fastq.gz files are located.</li>
<li>
<em>transcriptome (GUI Transcriptome folder)</em>, the path to the Cellranger compatible transcriptome reference</li>
<li>
<em>scratch.folder (GUI Scratch folder)</em>, a character string indicating the path of the scratch folder</li>
<li>
<em>expect.cells (GUI Expected cells)</em>, optional setting the number of recovered cells. Default: 3000 cells</li>
<li>
<em>version</em>, cell ranger version number, as default version 2 but version 3 is nowa also available.</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">home &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>()
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"/data/genomes/cellranger_hg19mm10"</span>)
<span class="co">#getting the human and mouse cellranger index</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://cf.10xgenomics.com/supp/cell-exp/refdata-cellranger-hg19-and-mm10-2.1.0.tar.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(home)
<span class="co"># downloading 100 cells 1:1 Mixture of Fresh Frozen Human (HEK293T) and Mouse (NIH3T3) Cells</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://cf.10xgenomics.com/samples/cell-exp/2.1.0/hgmm_100/hgmm_100_fastqs.tar"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"tar xvf hgmm_100_fastqs.tar"</span>)
<span class="co"># The cellranger analysis is run without the generation of the secondary analysis</span>
<span class="kw"><a href="../reference/cellrangerCount.html">cellrangerCount</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,  <span class="dt">transcriptome.folder=</span><span class="st">"/data/genomes/cellranger_hg19mm10"</span>,  
                <span class="dt">fastq.folder=</span><span class="st">"/data/test_cell_ranger/fastqs"</span>,  <span class="dt">expect.cells=</span><span class="dv">100</span>, 
                <span class="dt">nosecondary=</span><span class="ot">TRUE</span>, <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>, <span class="dt">version=</span><span class="st">"2"</span>)</code></pre></div>
<p></p>
<p>The analysis done above took 56.4 mins on a <a href="www.seqbox.com"><em>SeqBox</em></a>, equipped with an Intel i7-6770HQ (8 threads), 32 Gb RAM and 500Gb SSD.</p>
<p>The output of the above analysis are two counts matrices <strong>results_cellranger.cvs</strong> and <strong>results_cellranger.txt</strong> and a folder called <strong>results_cellranger</strong>, which contains the full cellranger output, more information on cellranger output can be found at <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/overview"><em>10XGenomics web site</em></a>.</p>
<p>In case a h5 matrix is available the function <strong>h5tocvs</strong> can convert the h5 format in csv and txt conventional matrices</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#download from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE106268 the GSE106268_RAW.tar file using a web browser</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"tar xvf GSE106268_RAW.tar"</span>)
<span class="kw"><a href="../reference/h5tocvs.html">h5tocvs</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"GSM2833284_Naive_WT_Rep1.h5"</span>,<span class="dt">sep=</span><span class="st">"/"</span>))</code></pre></div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help):
<ul>
<li>
<em>group</em>, a character string. Two options: sudo or docker, depending to which group the user belongs</li>
<li>
<em>file</em>, path of the h5 file. Full path MUST be included.</li>
</ul>
</li>
</ul>
</div>
</div>
<div id="section-3-counts-matrix-editing" class="section level2">
<h2 class="hasAnchor">
<a href="#section-3-counts-matrix-editing" class="anchor"></a><strong>Section 3</strong> Counts matrix editing</h2>
<p>This paragraph describes a set of functions that can be used to remove low quality cells and non-informative genes from counts tables generated by any single-cell sequencing platform.</p>
<ul>
<li>
<strong>Counts manipulation</strong>:
<ul>
<li>Removing non informative genes: <strong>filterZeros</strong>
</li>
<li>Plotting detected genes versus total number of UMIs (Unique Molecular Identifier)/reads: <strong>mitoRiboUmi</strong>
</li>
<li>Removing low quality cells: <strong>lorenzFilter</strong>
</li>
<li>Data normalization: <strong>scnorm</strong>, minimal requirements 10K counts/cell, works best with whole transcript sequencing</li>
<li>Data normalization: <strong>umiNorm</strong>, global normalization methods TMM and RLE are suitable for UMI data</li>
<li>ENSEMBL annotation and mitochodrial/ribosomal genes removal: <strong>scannobyGtf</strong>
</li>
<li>Converting a count table in log10: <strong>counts2log</strong>
</li>
<li>Removing cell cycle bias: <strong>recatPrediction</strong>/<strong>ccremove</strong>
</li>
</ul>
</li>
</ul>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui4.png" alt="GUI: Counts manipulation menu" width="20%"><p class="caption">
GUI: Counts manipulation menu
</p>
</div>
<div id="section-3-1-removing-non-informative-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-1-removing-non-informative-genes" class="anchor"></a><strong>Section 3.1</strong> Removing non informative genes</h3>
<p>The function <strong>filterZeros</strong> retains all genes having a user defined fraction of zeros (between 0 and 1, where 1 indicate that only genes without any 0s are retained, and 0 indicates that genes with at least a single value different from zero are retained), and plots the frequency distribution of gene counts in the dataset.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui5.png" alt="GUI: Filter Zeros panel" width="60%"><p class="caption">
GUI: Filter Zeros panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>file (GUI Counts table)</em>, a character string indicating the path of the tab delimited file of cells un-normalized expression counts</li>
<li>
<em>threshold (GUI Threshold)</em>, a number from 0 to 1 indicating the fraction of max accepted zeros in each gene. 0 is set as default and it eliminates only genes having no expression in any cell.</li>
<li>
<em>sep (GUI Separator)</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
</ul>
</li>
</ul>
<p>The output is a PDF providing zeros distributions before and after removal of genes with 0s counts. A tab delimited file with the prefix <strong>filtered_</strong> in which the filtered data are saved.</p>
<p><strong>IMPORTANT</strong>: In case user would like to apply cell quality filter, e.g. <strong>lorenzFilter</strong>, it is convenient to remove only genes with 0 counts in all cells, i.e. threshold=0 (Figure below). </p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/testSCumi_mm10.csv.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"testSCumi_mm10.csv.zip"</span>)
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"testSCumi_mm10.csv"</span>, <span class="dt">sep=</span><span class="st">","</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/dim">dim</a></span>(tmp)
<span class="co">#27998   806</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(tmp, <span class="st">"testSCumi_mm10.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names=</span><span class="ot">NA</span>)
<span class="kw"><a href="../reference/filterZeros.html">filterZeros</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">threshold=</span><span class="dv">0</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="co">#Out of 27998 genes 11255 are left after removing genes with no counts</span>
<span class="co">#output is filtered_testSCumi_mm10.txt</span></code></pre></div>
<p></p>
<div class="figure" style="text-align: center">
<img src="../inst/img/filterZero.jpeg" alt="Zeros distribution in full table, orange, and filtered table, blue" width="60%"><p class="caption">
Zeros distribution in full table, orange, and filtered table, blue
</p>
</div>

</div>
<div id="section-3-2-plotting-genes-numbers-versus-total-umisreads-in-each-cell" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-2-plotting-genes-numbers-versus-total-umisreads-in-each-cell" class="anchor"></a><strong>Section 3.2</strong> Plotting genes numbers versus total UMIs/reads in each cell</h3>
<p>To estimate the overall amount of genes detectable in each cell, the function <strong>genesUmi</strong> generates a plot of the number of genes present in a cell with respect to the total number of UMI in the same cell. The number of UMIs required to call a gene present in a cell is a parameter defined by the user, the suggested value is 3 UMIs. The function <strong>mitoRiboUmi</strong> plots <strong>genesUmi</strong> output and also generate the percentage of mitochondrial protein genes with respect to percentage of ribosomal protein genes fro each cell.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui6.png" alt="GUI: Genes versus counts panel" width="60%"><p class="caption">
GUI: Genes versus counts panel
</p>
</div>
<ul>
<li>
<em>Parameters genesUmi function</em> (only those without default; for the full list of parameters please refer to the function help):
<ul>
<li>
<em>file (GUI Counts table)</em>, a character string indicating the path of the file tab delimited of cells un-normalized expression counts.</li>
<li>
<em>umiXgene (UMIs X genes)</em>, an integer defining how many UMIs are required to call a gene present. default: 3</li>
<li>
<em>sep</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
</ul>
</li>
<li>
<em>Parameters mitoRiboUmi</em> (only those without default; for the full list of parameters please refer to the function help) (Figure ):
<ul>
<li>
<em>file (GUI Counts table)</em>, a character string indicating the path of the file tab delimited of cells un-normalized expression counts.</li>
<li>
<em>scratch.folder</em>, a character string indicating the path of the scratch folder</li>
<li>
<em>umiXgene (UMIs X genes)</em>, an integer defining how many UMIs are required to call a gene present. default: 3</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>gtf.name</em>, name for the gtf file to be used</li>
<li>
<em>bio.type</em>, ENSEMBL biotype of interest, default “protein_coding”</li>
</ul>
</li>
</ul>
<p>The output are two pdfs named respectively <strong>genes.umi.pdf</strong> (Figure below, panel A), where each dot represents a cell. X axis is the total number of UMIs/reads mapped on each cell in log10 format and Y axis is the number of detected genes and <strong>Ribo_mito.pdf</strong> (Figure below, panel B), Percentage of mitochondrial protein genes plotted with respect to percentage of ribosomal protein genes. Cells are colored on the basis of total number detected genes. The latter plot is useful to identify stressed cells, i.e. those with high percentage of mitochondrial genes (Figure below, panel B, black cells), and low informative cells, i.e. those in which genes are few and mainly ribosomal/mitochondrial (Figure below, panel B, black and green cells).</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget ftp://ftp.ensembl.org/pub/release-94/gtf/mus_musculus/Mus_musculus.GRCm38.94.gtf.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"gzip -d Mus_musculus.GRCm38.94.gtf.gz"</span>)
<span class="kw"><a href="../reference/mitoRiboUmi.html">mitoRiboUmi</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"filtered_testSCumi_mm10.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
            <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">umiXgene=</span><span class="dv">3</span>, 
            <span class="dt">gtf.name=</span><span class="st">"Mus_musculus.GRCm38.94.gtf"</span>, <span class="dt">bio.type=</span><span class="st">"protein_coding"</span>)

<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"filtered_testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</code></pre></div>
<p></p>
<p>In Figure  it is shown the distribution of genes in cells for ‘filtered_testSCumi_mm10.txt’ counts table.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/mitoRiboUmi.png" alt="mitoRiboUmi output: A) Number of detected genes plotted for each cell with respect to the total number of UMI/reads in that cell. B) Percentage of mitochondrial protein genes plotted with respect to percentage of ribosomal protein genes. Cells are colored on the basis of total number detected genes" width="100%"><p class="caption">
mitoRiboUmi output: A) Number of detected genes plotted for each cell with respect to the total number of UMI/reads in that cell. B) Percentage of mitochondrial protein genes plotted with respect to percentage of ribosomal protein genes. Cells are colored on the basis of total number detected genes
</p>
</div>

<div id="section-3-2-1-further-considerations-about-the-number-of-readsumis-to-be-used-in-a-single-cell-sequencing-experiment-" class="section level4">
<h4 class="hasAnchor">
<a href="#section-3-2-1-further-considerations-about-the-number-of-readsumis-to-be-used-in-a-single-cell-sequencing-experiment-" class="anchor"></a><strong>Section 3.2.1</strong> Further considerations about the number of reads/UMIs to be used in a single-cell sequencing experiment.</h4>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/28212749"><em>Ziegenhain et al.</em></a> published a comparison between single cell sequencing protocols and they show that, in a simulated experiment, at least 250K reads/cell are required for the detection of at least 80% of differentially expressed genes between two groups (Figure below). Ziegenhain observation clearly also apply to sub-populations clustering.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/Ziegenhain2017.jpeg" alt="Modified from Figure 6 in Ziegenhain et al. (Mol. Cell 2017). Power of scRNA-Seq methods. For more information on the experiment please see Ziegenhain paper." width="50%"><p class="caption">
Modified from Figure 6 in Ziegenhain et al. (Mol. Cell 2017). Power of scRNA-Seq methods. For more information on the experiment please see Ziegenhain paper.
</p>
</div>
<p>Sequencing depth, which affects differential expression analysis and sub-population partitioning, influences the structure of a single-cell dataset at two levels:</p>
<ul>
<li><p>number of genes called present in the experiment,</p></li>
<li><p>robustness of gene expression, i.e. number of reads associated to a gene.</p></li>
</ul>
<p>In particular, the number of genes called present in the experiment is the key element for the discrimination between sub-populations. In Figure below} it shown the effect of sequencing depth on the number of detectable genes in a set of 10XGenomics sequencing experiments (25K sequenced reads/cell extracted from CD19 B-cells [<a href="https://community.10xgenomics.com/t5/Data-Sharing/10x-Single-Cell-3-Paper-Zheng-et-al-2016-Datasets/td-p/231"><em>Zheng et al</em></a>], 83K sequenced reads/cell extracted from naive CD8+ T-cells [<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2833284"><em>GSM2833284</em></a>], and 250K sequenced reads/cell from an unpublished brain mouse experiment) and in an unpublished whole transcript human MAIT-cells single-cell sequencing done on Fluidigm C1 (25K, 100K, and 250K sequenced reads/cell were subsampled from the original fastqs). 3 UMI are used as minimal threshold to call present a gene in 10XGenomics experiments and 5 reads [<a href="https://genome.cshlp.org/content/21/12/2213.long"><em>Tarazona et al. 2011</em></a>] as minimal threshold to call a gene present in single cell whole transcriptome experiments.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#10XGenomics experiments</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/Zheng_cd19_288cells.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"Zheng_cd19_288cells.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"Zheng_cd19_288cells.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv genes.umi.pdf genes.umi_25k.pdf"</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/GSM2833284_Naive_WT_Rep1_288cell.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"GSM2833284_Naive_WT_Rep1_288cell.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"GSM2833284_Naive_WT_Rep1_288cell.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv genes.umi.pdf genes.umi_86k.pdf"</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/brain_unpublished_288cells.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"brain_unpublished_288cells.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"brain_unpublished_288cells.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv genes.umi.pdf genes.umi_250k.pdf"</span>)

<span class="co">#whole transcript experiment</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/c1_experiment.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"c1_experiment.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"c1_experiment"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"250K_counts.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">5</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv genes.umi.pdf genes.umi_250K.pdf"</span>)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"100K_counts.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">5</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv genes.umi.pdf genes.umi_100K.pdf"</span>)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"25K_counts.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">5</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv genes.umi.pdf genes.umi_25K.pdf"</span>)</code></pre></div>
<p></p>
<p>Figure below} clearly shows that the number of genes detectable by 10XGenomics sequencing (Figure below, panels A-C) is far less of those detectable using a whole transcript experiment (Figure below, panels D-F). In the case of 25K reads/cells in 3’ end sequencing (Figure A) the number of called genes goes from a doxen of genes to 350. In a whole transcript sequencing experiment where 25K reads/cells were considered (Figure below, panel D), 350 genes is the lowest number of genes detectable in a cell. In 10XGenomics experiment with 250K reads/cell the range of detectable genes goes from few hunderd to 2000, which is relatively similar to the number of detectable genes in a whole transcripts experiment where the same number of reads/cells were considered. The larger scattering and the overall lower number of detectable genes in 3’ end sequencing experiment, with respect to whole transcript experiments, is a peculiarity of the technology [<a href="https://www.ncbi.nlm.nih.gov/pubmed/28212749"><em>Ziegenhain et al.</em></a>].<br>
It has also to be highlighted that cells with a very low number of genes called present will have a gene repertoire made mainly of housekeeping genes, ribosomal and mitochondrial genes, see Figure  in <strong>Section 3.4</strong>. Thus, the lack of cell-type specific genes make these cells useless for the identification of functional cell sub-populations.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/seq_depth.jpeg" alt="Number of detected genes with respect to mapped reads. A) 25K reads/cell 10XGenomics platform, 3' end sequencing. B) 83K reads/cell 10XGenomics platform, 3' end sequencing. C) 250K reads/cell 10XGenomics platform, 3' end sequencing. D) 25K reads/cell C1 platform, whole transcript sequencing, E) 100K reads/cell C1 platform, whole transcript sequencing, F) 250K reads/cell C1 platform, whole transcript sequencing." width="100%"><p class="caption">
Number of detected genes with respect to mapped reads. A) 25K reads/cell 10XGenomics platform, 3’ end sequencing. B) 83K reads/cell 10XGenomics platform, 3’ end sequencing. C) 250K reads/cell 10XGenomics platform, 3’ end sequencing. D) 25K reads/cell C1 platform, whole transcript sequencing, E) 100K reads/cell C1 platform, whole transcript sequencing, F) 250K reads/cell C1 platform, whole transcript sequencing.
</p>
</div>
<p>The above observations indicate that 3’ end sequencing has a much lower genes called present with respect to whole transcript sequencing.</p>
<p>We have further investigated the following point:</p>
<ul>
<li>Is the number of total UMIs/cell affecting the separation between sub-populations?</li>
</ul>
<p>To address the above point we used two types of cells belonging to the T-cells [<a href="https://community.10xgenomics.com/t5/Data-Sharing/10x-Single-Cell-3-Paper-Zheng-et-al-2016-Datasets/td-p/231"><em>Zheng</em> 2016</a>]. The two sets of cells were sequenced with a coverage of approximately 21K reads/cell:</p>
<ul>
<li><p>T-cytotoxic (10209 cells, Figure A) cells,</p></li>
<li><p>T-regulatory (10263, Figure B) cells.</p></li>
</ul>
<p>We generated three datasets:</p>
<ul>
<li><p><strong>d3.4</strong>, which is made of 100 cells randomly selected within cells having, in each of the two datasets, a total UMIs/cell value greater than 2511 in each cell.</p></li>
<li><p><strong>d3</strong>, which is made of 100 cells randomly selected within cells having, in each of the two datasets, a total UMIs/cell value comprised between 2511 and 1000 in each cell.</p></li>
<li><p><strong>d3m</strong>, which is made of 100 cells randomly selected within cells having, in each of the two datasets, a total UMIs/cell smaller than 1000 in each cell.</p></li>
</ul>
<p>The separation between T-cytoxic and T-regulatory achievable with PCA is shown in Figure below, panels C-E.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/counts_effect.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(counts_effect.zip)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"counts_effect"</span>)

<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"df3.4.txt"</span>,<span class="dt">threshold=</span><span class="dv">1000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(docker4seq)
<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"df3.4_1000.txt"</span>, <span class="dt">type=</span><span class="st">"counts"</span>,
    <span class="dt">legend.position=</span><span class="st">"topleft"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
    <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>,
    <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())

<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"df3.txt"</span>,<span class="dt">threshold=</span><span class="dv">1000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(docker4seq)
<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"df3_1000.txt"</span>, <span class="dt">type=</span><span class="st">"counts"</span>,
    <span class="dt">legend.position=</span><span class="st">"topleft"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
    <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>,
    <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())

<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"df3m.txt"</span>,<span class="dt">threshold=</span><span class="dv">1000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(docker4seq)
<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"df3m_1000.txt"</span>, <span class="dt">type=</span><span class="st">"counts"</span>,
    <span class="dt">legend.position=</span><span class="st">"topleft"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
    <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>,
    <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())</code></pre></div>
<p></p>
<div class="figure" style="text-align: center">
<img src="../inst/img/counts_effect1.png" alt="Detectable genes in a Zheng  data subset. A) T-cytotoxic genes versus total cell reads plot. B) T-regulatory  genes versus total cell reads plot. C) d3m set, made of cells with less than 1000 counts each, D) df3, made of cells with counts between 1000 and 2511. E) d3.4, made of cells with more than 2511 counts each." width="100%"><p class="caption">
Detectable genes in a Zheng data subset. A) T-cytotoxic genes versus total cell reads plot. B) T-regulatory genes versus total cell reads plot. C) d3m set, made of cells with less than 1000 counts each, D) df3, made of cells with counts between 1000 and 2511. E) d3.4, made of cells with more than 2511 counts each.
</p>
</div>

<p>It is notable that only the datasets including cells with more than 2511 UMIs/cell (Figure above, panel E) is the one where the separation between the two T-cell types improves. In the d3 and d3m (Figure above, panels C,D) the amount of variance explained by the first component is much lower of that observable in d3.4 and the datasets are intersperse.</p>
<p>Thus, since 3’ end sequencing platforms (10Xgenomics, inDrop) has the advantage to produce high number of sequenced cells, users might decide to selected for clustering only the subset of cells with the highest number of genes called present.</p>
</div>
</div>
<div id="section-3-3-identifying-and-removing-cell-low-quality-outliers" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-3-identifying-and-removing-cell-low-quality-outliers" class="anchor"></a><strong>Section 3.3</strong> Identifying and removing cell low-quality outliers</h3>
<p>Lorenz statistics was implemented in rCASC <strong>lorenzFilter</strong> function. This function derives from the work of Diaz and coworkers [<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4937196/"><em>2016</em></a>] detecting low quality cells and this statistics correlates with live-dead staining [<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4937196/"><em>Diaz et al. 2016</em></a>].</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui7.png" alt="GUI: Lorenz Filter panel" width="60%"><p class="caption">
GUI: Lorenz Filter panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder (GUI Scratch folder)</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Counts table)</em>, full path to the count file MUST be provided</li>
<li>
<em>p_value (GUI p-value)</em>, lorenz statistics threshold, suggested value 0.05 (i.e. 5% probability that the a cell of low quality is selected)</li>
<li>
<em>separator (GUI Separator)</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
</ul>
</li>
</ul>
<p>The output is a counts table without low quality cells and with the prefix <strong>lorenz_filtered_</strong>. Output will be in the same format and with the same separator of input.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#example data set provided as part of rCASC package</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/testSCumi_mm10.csv.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"testSCumi_mm10.csv.zip"</span>)
<span class="co">#IMPORTANT: full path to the file MUST be cell count file included!</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="co"># the p_value indicate the probability that a low quality cell is retained in the </span>
<span class="co"># dataset filtered on the basis of Lorenz Statistics.</span>
<span class="kw"><a href="../reference/lorenzFilter.html">lorenzFilter</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>, 
               <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"filtered_testSCumi_mm10.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>),
               <span class="dt">p_value=</span><span class="fl">0.05</span>, <span class="dt">separator=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)

tmp0 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"filtered_testSCumi_mm10.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
<span class="co">#806 cells</span>

tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"lorenz_filtered_testSCumi_mm10.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
<span class="co">#785 cells</span></code></pre></div>
<p></p>
<p>In the example above 21 cells were removed because of their low quality (Figure below).</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/lorenz_filter.jpg" alt="Effect of Lorenz filtering, cells shown in blue have been discarded because of their low quality." width="40%"><p class="caption">
Effect of Lorenz filtering, cells shown in blue have been discarded because of their low quality.
</p>
</div>

</div>
<div id="section-3-4-annotation-and-mitocondrialribosomal-protein-genes-removal" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-4-annotation-and-mitocondrialribosomal-protein-genes-removal" class="anchor"></a><strong>Section 3.4</strong> Annotation and mitocondrial/ribosomal protein genes removal</h3>
<p>The function <strong>scannobyGtf</strong> allows the annotation of single-cell matrix, if ENSEMBL gene ids are provided. The function requires the ENSEMBL GTF of the organism under analysis and allows the selection of specific annotation biotypes, e.g. protein_coding:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui8.png" alt="GUI: Annotation panel" width="60%"><p class="caption">
GUI: Annotation panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>file (GUI Counts table)</em>, full path to the count file MUST be provided</li>
<li>
<em>gtf.name</em>, ENSEMBL gtf file name. GTF is located in the same folder where counts file is.</li>
<li>
<em>biotype</em>, biotype of interest. See <a href="https://www.ensembl.org/info/genome/genebuild/biotypes.html"><em>www.ensembl.org/info/genome/genebuild/biotypes.html</em></a> for more information</li>
<li>
<em>mt (GUI Mt)</em>, a boolean to define if mitochondrial genes have to be removed, FALSE mean that mt genes are removed</li>
<li>
<em>ribo.proteins (GUI Ribo Proteins)</em>, a boolean to define if ribosomal proteins have to be removed, FALSE mean that ribosomal proteins (gene names starting with rpl or rps) are removed</li>
<li>
<em>umiXgene (GUI UMIs X gene)</em>, a integer defining how many UMIs are required to call a gene as present. default: 3</li>
<li>
<em>riboStart.percentage</em>, start range for ribosomal percentage, cells within the range are kept</li>
<li>
<em>riboEnd.percentage</em>, end range for ribosomal percentagem cells within the range are kept</li>
<li>
<em>mitoStart.percentage</em>, start range for mitochondrial percentage, cells within the range are retained</li>
<li>
<em>mitoEnd.percentage</em>, end range for mitochondrial percentage, cells within the range are retained</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#running annotation and removal of mito and ribo proteins genes</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget ftp://ftp.ensembl.org/pub/release-92/gtf/mus_musculus/Mus_musculus.GRCm38.94.gtf.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"gunzip Mus_musculus.GRCm38.94.gtf.gz"</span>)
<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"lorenz_filtered_testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"Mus_musculus.GRCm38.94.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">TRUE</span>, <span class="dt">ribo.proteins=</span><span class="ot">TRUE</span>, <span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">riboStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">riboEnd.percentage=</span><span class="dv">100</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, <span class="dt">mitoEnd.percentage=</span><span class="dv">100</span>)

<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"annotated_lorenz_filtered_testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"filtered_annotated_lorenz_filtered_testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</code></pre></div>
<p></p>
<p>The output are a file with prefix <strong>annotated_</strong>, containing all annotated genes and a file with prefix <strong>filtered_annotated_</strong> wich contain the filtered subset of cells.</p>
<p>Ribosomal RNA and ribosomal proteins represent a significant part of cell cargo. Large cells and actively proliferating cells will have respectively more ribosomes and more active ribosome synthesis [<a href="https://www.ncbi.nlm.nih.gov/pubmed/18583314"><em>Montanaro et al. 2008</em></a>]. Thus, ribosomal proteins expression might represent one of the major confounding factor in cluster formation between active and dormient cells. Furthermore, the main function of mitochondria is to produce energy through aerobic respiration. The number of cell mitochondria depends on its metabolic demands [<a href="https://www.ncbi.nlm.nih.gov/pubmed/25200564"><em>Nasrallah and Horvath 2014</em></a>]. This might also affect clustering, favoring the separation between metabolic active and resting cells, with respect to functional differences between sub-populations. <em>scannobyGtf</em> allows also the removal of mitocondrial and ribosomal protein genes. </p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"lorenz_filtered_testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"Mus_musculus.GRCm38.94.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">FALSE</span>, <span class="dt">ribo.proteins=</span><span class="ot">FALSE</span>,<span class="dt">umiXgene=</span><span class="dv">3</span>,
                    <span class="dt">riboStart.percentage=</span><span class="dv">10</span>, <span class="dt">riboEnd.percentage=</span><span class="dv">70</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">mitoEnd.percentage=</span><span class="dv">5</span>)

<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"annotated_lorenz_filtered_testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"filtered_annotated_lorenz_filtered_testSCumi_mm10.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</code></pre></div>
<p></p>
<p>In figure below, panel B is shown the effect of the removal of both mitochondrial and ribosomal protein genes and the removal of cells characterized by high percentage of mtochondrial protein genes (&gt;5%) and too litte (&lt;10%) or too much (&gt;70%) ribosomal protein genes .</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/noMT-rib.png" alt="Removing mitochondrial and ribosomal proteins genes. A) All annotated cells. B) Only cells passing the filters" width="100%"><p class="caption">
Removing mitochondrial and ribosomal proteins genes. A) All annotated cells. B) Only cells passing the filters
</p>
</div>
</div>
<div id="section-3-5-top-expressed-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-5-top-expressed-genes" class="anchor"></a><strong>Section 3.5</strong> Top expressed genes</h3>
<p>For clustering purposes user might decide to use the top expressed genes. The function <strong>topx</strong> select the X top expressed genes given a user defined threshold. The function also produces a pdf file gene_expression_distribution.pdf showing the changes in the UMIs/gene expression distribution upon <strong>topx</strong> filtering:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui9.png" alt="GUI: Top X expressed genes panel" width="60%"><p class="caption">
GUI: Top X expressed genes panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>file (GUI Counts table)</em>, full path to the count file MUST be provided</li>
<li>
<em>threshold (GUI Top expressed genes)</em>, number of top expressed genes to be selected. Default 0, i.e. only genes will all cell values equal to 0 will be removed</li>
<li>
<em>logged (GUI Logged)</em>, boolean, if FALSE gene expression data are log10 transformed before being plotted.</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="dt">counts.matrix=</span><span class="st">"lorenz_filtered_testSCumi_mm10.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)
<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"lorenz_filtered_testSCumi_mm10.txt, sep="</span><span class="op">/</span><span class="st">")"</span>,<span class="dt">threshold=</span><span class="dv">10000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
<span class="kw"><a href="../reference/genesUmi.html">genesUmi</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="dt">counts.matrix=</span><span class="st">"lorenz_filtered_testSCumi_mm10_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">umiXgene=</span><span class="dv">3</span>,  <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</code></pre></div>
<p></p>
<p><strong>IMPORTANT</strong>: The core clustering tool in rCASC is SIMLR, <strong>Section 5</strong>. SIMLR requires that the number of genes must be larger then the number of analysed cells.</p>
</div>
<div id="section-3-6-data-normalization" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-6-data-normalization" class="anchor"></a><strong>Section 3.6</strong> Data normalization</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5549838/"><em>The best way to normalize single-cell RNA-seq data has not yet been resolved</em></a>, expecially in the case of UMI data. We inserted in our workflow two possible options:</p>
<ul>
<li><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5473255/"><em>SCnorm</em></a>, which works best with whole transcript data.</p></li>
<li><p><a href="https://www.biorxiv.org/content/early/2017/12/16/235382"><em>scone</em></a>, which provides different global scaling methods that can be applyed to UMI single-cell data.</p></li>
</ul>
<div id="section-3-6-1-scnorm" class="section level4">
<h4 class="hasAnchor">
<a href="#section-3-6-1-scnorm" class="anchor"></a><strong>Section 3.6.1</strong> SCnorm</h4>
<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5473255/"><em>SCnorm</em></a> performs a quantile-regression based approach for robust normalization of single-cell RNA-seq data. SCnorm groups genes based on their count-depth relationship then applies a quantile regression to each group in order to estimate scaling factors which will remove the effect of sequencing depth from the counts.</p>
<p>IMPORTANT: SCnorm is not intended for datasets with more than ~80% zero counts, because of lack of algoritm convergence in these situations.</p>
<div id="section-3-6-1-1-check-counts-depth-relationship" class="section level5">
<h5 class="hasAnchor">
<a href="#section-3-6-1-1-check-counts-depth-relationship" class="anchor"></a><strong>Section 3.6.1.1</strong> Check counts-depth relationship</h5>
<p>Before normalizing using <strong>scnorm</strong>, it is advised to check the data count-depth relationship. If all genes have a similar relationship then a global normalization strategy such as median-by-ratio in the DESeq package or TMM in edgeR will also be adequate. However, when the count-depth relationship varies among genes global scaling strategies leads to poor normalization. In these cases the normalization provided by SCnorm is recommended:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui10.png" alt="GUI: SCnorm: check counts-depth relationship panel" width="60%"><p class="caption">
GUI: SCnorm: check counts-depth relationship panel
</p>
</div>
<p><strong>checkCountDepth</strong> provides a wrapper, in rCASC, for the checkCountDepth of the <a href="https://github.com/rhondabacher/SCnorm"><em>SCnorm package</em></a>, which estimates the count-depth relationship for all genes.</p>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>file (GUI Counts table)</em>, full path to the file MUST be included. Only tab delimited files are supported</li>
<li>
<em>conditions</em>, vector of condition labels, this should correspond to the columns of the un-normalized expression matrix. If not provided data is assumed to come from same condition/batch.</li>
<li>
<em>ditherCounts (GUI UMI)</em>, boolean. Setting to TRUE might improve results with UMI data.</li>
<li>
<em>FilterCellProportion (GUI Min non-zero cells)</em>, a value indicating the proportion of non-zero expression estimates required to include the genes into the evaluation. Default is .10, and will not go below a proportion which uses less than 10 total cells/samples</li>
<li>
<em>FilterExpression (GUI Med non-zero expr. threshold)</em>, a value indicating exclude genes having median of non-zero expression below this threshold from count-depth plots #‘<span class="citation">@param</span> ditherCounts, Setting to TRUE might improve results with UMI data, default is FALSE #’ <span class="citation">@param</span> outputName, specify the path and/or name of output files.</li>
<li>
<em>outputName</em>, name of output files.</li>
<li>
<em>nCores</em>, number of cores to use.</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#this specific example is an UMI counts table made of 12 cells having at least 10K UMIs/cell.</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/example_UMI.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"example_UMI.txt.zip"</span>)
conditions=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span>,<span class="dv">12</span>)
<span class="kw"><a href="../reference/checkCountDepth.html">checkCountDepth</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"example_UMI.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>),
     <span class="dt">conditions=</span>conditions, <span class="dt">FilterCellProportion=</span><span class="fl">0.1</span>, <span class="dt">FilterExpression=</span><span class="dv">0</span>,
     <span class="dt">ditherCounts=</span><span class="ot">TRUE</span>, <span class="dt">outputName=</span><span class="st">"example_UMI"</span>, <span class="dt">nCores=</span><span class="dv">8</span>)</code></pre></div>
<p></p>
<p>The output is a PDF (Figure below), providing a view of the counts distribution, and a file selected.genes.txt, which contains the genes selected to run the analysis.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/checkCountDepth0.jpg" alt="checkCountDepth output plot" width="40%"><p class="caption">
checkCountDepth output plot
</p>
</div>

</div>
<div id="section-3-6-1-2-scnorm" class="section level5">
<h5 class="hasAnchor">
<a href="#section-3-6-1-2-scnorm" class="anchor"></a><strong>Section 3.6.1.2</strong> scnorm</h5>
<p>The <strong>scnorm</strong> function executes SCnorm from <a href="https://github.com/rhondabacher/SCnorm"><em>SCnorm package</em></a>, which normalizes across cells to remove the effect of sequencing depth on the counts and returns the normalized expression count:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui11.png" alt="GUI: SCnorm: counts-depth normalization panel" width="70%"><p class="caption">
GUI: SCnorm: counts-depth normalization panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>file (GUI Counts table)</em>, full path to the file MUST be included. Only tab delimited files are supported</li>
<li>
<em>conditions</em>, vector of condition labels, this should correspond to the columns of the un-normalized expression matrix.</li>
<li>
<em>outputName</em>, specify the name of output files.</li>
<li>
<em>nCores</em>, number of cores to use.</li>
<li>
<em>filtercellNum (Min non-zero cells)</em>, the number of non-zero expression estimate required to include the genes into the SCnorm fitting (default = 10). The initial grouping fits a quantile regression to each gene, making this value too low gives unstable fits.</li>
<li>
<em>ditherCount (GUI UMI)</em>, boolean. Setting to TRUE might improve results with UMI data.</li>
<li>
<em>FilterExpression (GUI Med non-zero expr. threshold)</em>, a value indicating exclude genes having median of non-zero expression below this threshold from count-depth plots</li>
<li>
<em>PropToUse (GUI Prop. to use)</em>, as default is set to 0.25, but to increase speed with large data set could be reduced, e.g. 0.1</li>
<li>
<em>PrintProgressPlots</em>, boolean. If it is set to TRUE produces a plot as SCnorm determines the optimal number of groups</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/example_UMI.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"example_UMI.txt.zip"</span>)
<span class="co">#this specific example is an UMI counts table made of 12 cells having at least 10K UMIs/cell.</span>
conditions=<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span>,<span class="dv">12</span>)
<span class="kw"><a href="../reference/scnorm.html">scnorm</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"example_UMI.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>),
     <span class="dt">conditions=</span>conditions,<span class="dt">outputName=</span><span class="st">"example_UMI"</span>, <span class="dt">nCores=</span><span class="dv">8</span>, <span class="dt">filtercellNum=</span><span class="dv">10</span>,
     <span class="dt">ditherCount=</span><span class="ot">TRUE</span>, <span class="dt">PropToUse=</span><span class="fl">0.1</span>, <span class="dt">PrintProgressPlots=</span><span class="ot">TRUE</span>, <span class="dt">FilterExpression=</span><span class="dv">1</span>)</code></pre></div>
<p></p>
<p>The output files are plots of the normalization effects (Figure below), a tab delimited file containing the normalized data, with the prefix <strong>normalized_</strong>, and <strong>discarded_genes.txt</strong>, which contains the discarded genes.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/scnorm.jpeg" alt="Effect of the SCnorm on the dataset in Figure 9" width="40%"><p class="caption">
Effect of the SCnorm on the dataset in Figure 9
</p>
</div>
<p><strong>scnorm</strong> is compliant with <a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>SIMLR</em></a>, the rCASC core clustering tool.</p>
</div>
</div>
<div id="section-3-6-2-scone" class="section level4">
<h4 class="hasAnchor">
<a href="#section-3-6-2-scone" class="anchor"></a><strong>Section 3.6.2</strong> scone</h4>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui12.png" alt="GUI: Normalization panel" width="70%"><p class="caption">
GUI: Normalization panel
</p>
</div>
<p>scone package embeds:</p>
<ul>
<li><p>Centered log-ratio (<strong>CLR</strong>) normalization</p></li>
<li>
<p>Relative log-expression (<strong>RLE</strong>; DESeq) scaling normalization</p>
<ul>
<li>the scaling factors are calculated for each lane as median of the ratio, for each gene, of its read count of its geometric mean across all lanes.</li>
</ul>
</li>
<li>
<p>Full-quantile normalization</p>
<ul>
<li>quantile normalization is a technique for making two or more distributions identical in statistical properties. To quantile normalize two or more samples to each other, sort the samples, then set to the average (usually, arithmetic mean) of the samples. So the highest value in all cases becomes the mean of the highest values, the second highest value becomes the mean of the second highest values, and so on.</li>
</ul>
</li>
<li><p>Simple deconvolution normalization</p></li>
<li>
<p>Sum scaling normalization</p>
<ul>
<li>Gene counts are divided by the total number of mapped reads (or library size) associated with their lane and multiplied by the mean total count across all the samples of the dataset.</li>
</ul>
</li>
<li>
<p>Weighted trimmed mean of M-values (<strong>TMM</strong>, edgeR) scaling normalization (suitable for single-cell)</p>
<ul>
<li>to compute the TMM factor, one lane is considered a reference sample and the others test samples, with TMM being the weighted mean of log ratios between test and reference, after excluding the most expressed genes and the genes with the largest log ratios.</li>
</ul>
</li>
<li>
<p>Upper-quartile (<strong>UQ</strong>) scaling normalization</p>
<ul>
<li>the total counts are replaced by the upper quartile of counts different from 0 in the computation of the normalization factors.</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Weighted trimmed mean of M-values (TMM) scaling normalization</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/example_UMI.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"example_UMI.txt.zip"</span>)
<span class="kw"><a href="../reference/umiNorm.html">umiNorm</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"example_UMI.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>),
       <span class="dt">outputName=</span><span class="st">"example_UMI"</span>, <span class="dt">normMethod=</span><span class="st">"TMM_FN"</span>)</code></pre></div>
<p></p>
<p><strong>IMPORTANT</strong>: In case sub-population discovery is the analysis task, it is important to check if a specific normalization is compliant with the clustering approach in use. For example in the case of <a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>SIMLR</em></a>, the rCASC core clustering tool, the normalizations provided in <strong>scone</strong> are not compliant, because they remove some of the features required to run the SIMLR multikernel learning analysis. TMM is instead compliant with the rCASC implementation of <a href="https://lvdmaaten.github.io/tsne/"><strong>tSne</strong></a>. In case <a href="https://satijalab.org/seurat/pbmc3k_tutorial.html"><em>Seurat</em></a> clustering is used the data set does not required any normalization since a normalization procedure is included in the algorithm.</p>
</div>
</div>
<div id="section-3-7-log-conversion-of-a-count-table" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-7-log-conversion-of-a-count-table" class="anchor"></a><strong>Section 3.7</strong> Log conversion of a count table</h3>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui13.png" alt="GUI: Log transformation for counts table panel" width="70%"><p class="caption">
GUI: Log transformation for counts table panel
</p>
</div>
<p>The function <strong>counts2log</strong> can convert a count table in a log10 values saved in a comma separated or tab delimited file.</p>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>file (GUI Counts table)</em>, full path to the file MUST be included.</li>
<li>
<em>log.base</em>, the base of the log to be used for the transformation</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/counts2log.html">counts2log</a></span>(<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"example_UMI.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">log.base=</span><span class="dv">10</span>)</code></pre></div>
<p></p>
</div>
<div id="section-3-8-detecting-and-removing-cell-cycle-bias" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-8-detecting-and-removing-cell-cycle-bias" class="anchor"></a><strong>Section 3.8</strong> Detecting and removing cell cycle bias</h3>
<p>Single-cell RNA-Sequencing measurement of expression often suffers from large systematic bias. A major source of this bias is cell cycle, which introduces large within-cell-type heterogeneity that can obscure the differences in expression between cell types. <a href="https://www.nature.com/articles/srep33892"><em>Barron and Li</em></a> developed in 2016 a R package called <a href="https://cran.r-project.org/web/packages/ccRemover/index.html"><em>ccRemover</em></a> which removes cell cycle effects and preserves other biological signals of interest.</p>
<p>However, before applying <em>ccRemover</em>, it is essential to address if the removal of cell cycle effect is required. <a href="https://www.nature.com/articles/s41467-017-00039-z"><em>reCAT</em></a> is a modeling framework for unsynchronized single-cell transcriptome data that can reconstruct cell cycle time-series. Thus, reCAT cell cycle prediction step can be used to check if cell cycle effect can be detected in a dataset and therefore ccRemover normalization approach will be needed.</p>
<div id="section-3-8-1-evaluating-the-presence-of-cell-cycle-effect-in-a-dataset-recat" class="section level4">
<h4 class="hasAnchor">
<a href="#section-3-8-1-evaluating-the-presence-of-cell-cycle-effect-in-a-dataset-recat" class="anchor"></a><strong>Section 3.8.1 </strong> Evaluating the presence of cell cycle effect in a dataset: <a href="https://www.nature.com/articles/s41467-017-00039-z"><em>reCAT</em></a>
</h4>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui14.png" alt="GUI: cell cycle estimation panel" width="70%"><p class="caption">
GUI: cell cycle estimation panel
</p>
</div>
<p><strong>reCAT</strong> prediction step is implemented in rCASC in the function <strong>recatPrediction</strong>, which requires a data set annotated using <strong>scannobyGtf</strong>.</p>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Counts table)</em>, the path to the input file</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>geneNameControl (GUI gene name format)</em>, 0 if the matrix has gene symbol without ENSEMBL code. 1 if the gene names is formatted like this : ENSMUSG00000000001:Gnai3. If the gene names is only ENSEMBL name SCannoByGtf has to be executed before recatPrediction.</li>
<li>
<em>seed</em>, important parameter for reproduce the same result with the same input, default 111</li>
<li>
<em>window (GUI N. cells/point)</em>, number of cell plotted per point</li>
</ul>
</li>
</ul>
<p>To show the differences existing between a dataset characterized by cell cycle bias and one that is not, we used two datasets:</p>
<ul>
<li><p>the dataset published by [<a href="https://www.ncbi.nlm.nih.gov/pubmed/25599176"><em>Buettner et al.</em></a>], containing naive-T-cells and T-helper2-cells mixed together and sorted on the basis of the cell cycle state.</p></li>
<li><p>The quiescent naive T-cell dataset part of the publication of <a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=29326266"><em>Pace et al.</em></a>, expected to be in G0.</p></li>
</ul>
<p>To execute the analysis on the same number of cells, 288 cells were randomly selected from quiescent naive T-cell dataset. In Figure below, panel A the presence of oscillatory behaviour is evident in the predicted cells time serie and the G1 and G2M trends are indicated respectively in blue and red dashed curves. On the other hand the oscillatory behaviour is totally absent (Figure below, panel B) in the naive T-cells, which are expected to be quiescent in G0.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/recat1.jpeg" alt="Cell cycle assignment to the cells. A) Buettner et al. (Nat. Biotechnol. 2015) raw dataset, cells are expected to be distributed in G1, S and G2M, B) Naive T-cells, expected to be mainly in G0 (Science 2018)." width="70%"><p class="caption">
Cell cycle assignment to the cells. A) Buettner et al. (Nat. Biotechnol. 2015) raw dataset, cells are expected to be distributed in G1, S and G2M, B) Naive T-cells, expected to be mainly in G0 (Science 2018).
</p>
</div>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#preparing the data for the analysis</span>
 <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/buettner_G1G2MS_counts.txt.zip"</span>)
 <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"buettner_G1G2MS_counts.txt.zip"</span>)

<span class="co">#annotatiing the data set to obtain the gene names in the format ensemblID:symbol </span>
<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"buettner_G1G2MS_counts.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"Mus_musculus.GRCm38.92.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">TRUE</span>, <span class="dt">ribo.proteins=</span><span class="ot">TRUE</span>,<span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">riboStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">riboEnd.percentage=</span><span class="dv">100</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, <span class="dt">mitoEnd.percentage=</span><span class="dv">100</span>)

<span class="co">#selecting the top 10000 most expressed genes</span>
<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"annotated_buettner_G1G2MS_counts.txt"</span>,<span class="dt">threshold=</span><span class="dv">10000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)

<span class="co">#running cell cycle prediction</span>
<span class="kw"><a href="../reference/recatPrediction.html">recatPrediction</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>,
                <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_buettner_G1G2MS_counts_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">geneNameControl=</span><span class="dv">1</span>, <span class="dt">window=</span><span class="dv">10</span>, <span class="dt">seed=</span><span class="dv">111</span>)


<span class="co">#same analysis as above on 10XGenomix data of quescent naive-T cells</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/GSM2833284_Naive_WT_Rep1_288cell.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"GSM2833284_Naive_WT_Rep1_288cell.txt.zip"</span>)
<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"GSM2833284_Naive_WT_Rep1_288cell.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"Mus_musculus.GRCm38.92.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">TRUE</span>, <span class="dt">ribo.proteins=</span><span class="ot">TRUE</span>,<span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">riboStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">riboEnd.percentage=</span><span class="dv">100</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, <span class="dt">mitoEnd.percentage=</span><span class="dv">100</span>)
<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"annotated_GSM2833284_Naive_WT_Rep1_288cell.txt"</span>,<span class="dt">threshold=</span><span class="dv">10000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
<span class="kw"><a href="../reference/recatPrediction.html">recatPrediction</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>,
                <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_GSM2833284_Naive_WT_Rep1_288cell_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">geneNameControl=</span><span class="dv">1</span>, <span class="dt">window=</span><span class="dv">10</span>, <span class="dt">seed=</span><span class="dv">111</span>)</code></pre></div>
<p></p>
</div>
<div id="section-3-8-2-removing-cell-cycle-effect-in-a-dataset-ccremover" class="section level4">
<h4 class="hasAnchor">
<a href="#section-3-8-2-removing-cell-cycle-effect-in-a-dataset-ccremover" class="anchor"></a><strong>Section 3.8.2 </strong> Removing cell cycle effect in a dataset: <a href="https://cran.r-project.org/web/packages/ccRemover/index.html"><em>ccRemover</em></a>
</h4>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui15.png" alt="GUI: cell cycle bias removal panel" width="70%"><p class="caption">
GUI: cell cycle bias removal panel
</p>
</div>
<p><strong>ccRemover</strong> software is implemented in rCASC in the function <strong>ccRemove</strong>, which also requires a data set annotated using <strong>scannobyGtf</strong>.</p>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Counts table)</em>, the path to the input file, including the counts table name</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>seed</em>, is important to reproduce the same results with the same input, default=111</li>
<li>
<em>cutoff</em>, p-value to use: 3 is almost equal to 0.05</li>
<li>
<em>species</em>, human or mouse</li>
<li>
<em>rawCount</em>, 1 for unlogged and not-normalized, 0 otherwise</li>
</ul>
</li>
</ul>
<p><strong>IMPORTANT</strong>: The output of ccRemover does not require log transformation before clustering analysis.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#removing cell cycle effect</span>
<span class="kw"><a href="../reference/ccRemove.html">ccRemove</a></span>(<span class="dt">group=</span><span class="st">"docker"</span> , <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>,
        <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"annotated_buettner_G1G2MS_counts_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,
        <span class="dt">seed=</span><span class="dv">111</span>, <span class="dt">cutoff=</span><span class="dv">3</span>, <span class="dt">species=</span><span class="st">"mouse"</span>, <span class="dt">rawCount=</span><span class="dv">1</span>)</code></pre></div>
<p></p>
<p>The analyses above were done using a <a href="www.seqbox.com">SeqBox</a>, equipped with an Intel i7-6770HQ (8 threads), 32 GB RAM and 500 GB SSD. They took 54 and 40 mins for recatPrediction respectively on Buettner and the naive T-cell datasets. 28 mins were needed by ccRemove on Buettner data set. ccRemover analysis produces a ready-for-clustering data normalized matrix. The matrix can be identify by the prefix <strong>LS_cc_</strong>. <strong>ccRemove</strong> output is compliant with <a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>SIMLR</em></a>, the rCASC core clustering tool. <strong>ccRemove</strong> output does not require log transformation when applied to SIMLR.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#visualizing the dataset before and after cell cycle bias removal</span>
<span class="co">#reformat the matrix header to be suitable with docker4seq PCA plotting function</span>
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"annotated_buettner_G1G2MS_counts_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
tmp.n1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">1</span>])
tmp.n2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">2</span>])
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(tmp.n2, tmp.n1, <span class="dt">sep=</span><span class="st">"_"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(tmp, <span class="st">"annotated_buettner_G1G2MS_counts_10000bis.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names=</span><span class="ot">NA</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(devtools)
<span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"kendomaniac/docker4seq"</span>, <span class="dt">ref=</span><span class="st">"master"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(docker4seq)

<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"annotated_buettner_G1G2MS_counts_10000bis.txt"</span>, <span class="dt">type=</span><span class="st">"counts"</span>, 
      <span class="dt">legend.position=</span><span class="st">"topright"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
      <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>, 
      <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())

<span class="co">#reformat the matrix header to be suitable with docker4seq PCA plotting function</span>
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"LS_cc_annotated_buettner_G1G2MS_counts_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
tmp.n1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">1</span>])
tmp.n2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">2</span>])
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(tmp.n2, tmp.n1, <span class="dt">sep=</span><span class="st">"_"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(tmp, <span class="st">"LS_cc_annotated_buettner_G1G2MS_counts_10000bis.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names=</span><span class="ot">NA</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"LS_cc_annotated_buettner_G1G2MS_counts_10000bis.txt"</span>, <span class="dt">type=</span><span class="st">"TPM"</span>, 
      <span class="dt">legend.position=</span><span class="st">"topright"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
      <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>, 
      <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())</code></pre></div>
<p></p>
<p>In Figure below} are shown the results obtained using the ccRemove implementation in rCASC, using the Buettner dataset. The removal of the cell cycle effect (Figure below, panel B) is clearly shown by a reduction of the variance explained by PC1 and PC2 in the PCA plot.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/buettner_raw_ccremover.jpeg" alt="rCASC implementation of the ccRemove. A) PCA analysis of Buettner et al. (Nat. Biotechnol. 2015) raw dataset, B) PCA analysis of ccRemove cell-cycle normalized dataset." width="70%"><p class="caption">
rCASC implementation of the ccRemove. A) PCA analysis of Buettner et al. (Nat. Biotechnol. 2015) raw dataset, B) PCA analysis of ccRemove cell-cycle normalized dataset.
</p>
</div>
</div>
</div>
</div>
<div id="section-4-estimating-the-number-of-cluster-to-be-used-for-cell-sub-population-discovery-" class="section level2">
<h2 class="hasAnchor">
<a href="#section-4-estimating-the-number-of-cluster-to-be-used-for-cell-sub-population-discovery-" class="anchor"></a><strong>Section 4</strong> Estimating the number of cluster to be used for cell sub-population discovery.</h2>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui16.png" alt="GUI: Clustering panel" width="30%"><p class="caption">
GUI: Clustering panel
</p>
</div>
<p>The rCASC core clustering tool is <a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>SIMLR</em></a>, which requires as input the number of clusters to be used to aggregate cell sub-populations. Unfortunately, there is no definitive answer to the definition of the most probable number of clusters, in which cells will aggregate. Some of the possible ways to identify the most probable number of clusters is summaries in: <a href="http://www.sthda.com/english/wiki/print.php?id=239"><em>“Determining the optimal number of clusters: 3 must known methods - Unsupervised Machine Learning”</em></a>.</p>
<p>Another important aspect is how the number of detectable clusters might change if the number of cells changes in the dataset, e.g. upon removal of a random subset of cells. Because single-cell experiment, at least today, are rarely characterized by biological replications and frequently they represent the initial step of an analysis aimed at the identification for new cell sub-populations, it is very important to assess the stability of cells aggregations detected by clustering methods.</p>
<div id="section-4-1-estimating-the-number-of-cluster-to-be-used-for-cell-sub-population-discovery-by-community-detection-method-" class="section level3">
<h3 class="hasAnchor">
<a href="#section-4-1-estimating-the-number-of-cluster-to-be-used-for-cell-sub-population-discovery-by-community-detection-method-" class="anchor"></a><strong>Section 4.1</strong> Estimating the number of cluster to be used for cell sub-population discovery by community detection method.</h3>
<p>In rCASC, the identification of the optimal number of clusters is addressed, in presence of cells number perturbations, with <a href="https://ppapasaikas.github.io/griph/"><em>griph</em></a>. The clustering performed by griph is graph-based and uses the community detection method <a href="http://igraph.org/r/doc/cluster_louvain.html"><em>louvain modularity</em></a>. Griph algorithm is closer to agglomerative clustering methods, since every node is initially assigned to its own community and communities are subsequently built by iterative merging. Griph is embedded <strong>clusterNgriph</strong> function, which evaluates the number of clusters in which a set of cells will aggregate upon a user defined leave-N%-out cells bootstraps. In the example below the number of clusters are detected for the file ‘annotated_buettner_G1G2MS_counts_10000bis.txt’, used in <strong>Section 3.8</strong>.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui17.png" alt="GUI: Range of numbers of clusters estimation panel" width="70%"><p class="caption">
GUI: Range of numbers of clusters estimation panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Matrix counts)</em>, the path to the input file, including the counts table name</li>
<li>
<em>nPerm</em>, number of permutations to perform</li>
<li>
<em>permAtTime</em>, number of permutations that can be computed in parallel</li>
<li>
<em>percent</em>, percentage of randomly selected cells removed in each permutation</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>logTen</em>, integer, 1 if the count matrix is already in log10, 0 otherwise</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input, default is 111</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="kw"><a href="../reference/clusterNgriph.html">clusterNgriph</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), 
              <span class="st">"annotated_buettner_G1G2MS_counts_10000bis.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nPerm=</span><span class="dv">160</span>, 
              <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)</code></pre></div>
<p></p>
<p>In Figure below} it is shown the output generated by <strong>clusterNgriph</strong>. The output folder is called <strong>Results</strong> and it is located in the folder from which the analysis started. Within Results is present a folder named as the dataset used for the analysis. In this case ‘annotated_buettner_G1G2MS_counts_10000bis’. In the latter folder is present a folder, in this specific example ‘5’, named with the number of clusters that were more represented as result of the bootstrap analysis. The file indicated with the blue arrow contains all the information to generate the griph output plot, used as reference to allocate cells to a specific cluster at each bootstrap step. The file indicated with the green arrow contains the cluster position for each cell over all bootstrap steps. The file indicated with the red arrow contains the cells removed at each bootstrap steps. The file called ‘hist.pdf’, indicated with the black arrow, is the plot of the frequency of different number of clusters generated by griph as consequence of the bootstrap steps. In this specific case, over 160 permutations, 80 produced 5 clusters, 70 produced 4 clusters and 10 produced 6 clusters.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/griph3.jpeg" alt="Output of clusterNgriph" width="100%"><p class="caption">
Output of clusterNgriph
</p>
</div>
<p>It has to be noted that in principle, since this dataset has a strong cell cycle effect, we would have expected ideally only three clusters: G1, S and G2M. This toy experiment clearly shows that perturbation of the dataset under analysis can affect the number of detectable clusters. Thus, to identify the clustering condition that guarantee the greatest cell stability in a cluster, in our opinion it is mandatory clustering cells taking in account perturbation effects. In Section 4.2 we further investigate this issue.</p>
<div id="section-4-2-k-mean-clustering-investigating-how-cell-sub-populations-aggregation-is-affected-by-dataset-perturbations-" class="section level4">
<h4 class="hasAnchor">
<a href="#section-4-2-k-mean-clustering-investigating-how-cell-sub-populations-aggregation-is-affected-by-dataset-perturbations-" class="anchor"></a><strong>Section 4.2</strong> K-mean clustering: Investigating how cell sub-populations aggregation is affected by dataset perturbations.</h4>
<p>As indicated above we are interested to identify not only the optimal cluster number but also if the cluster number is affected by removal of a random subset of cells. To observe the effect of datasets perturbation in clustering we built 4 datasets combining different cell types available in <a href="https://community.10xgenomics.com/t5/Data-Sharing/10x-Single-Cell-3-Paper-Zheng-et-al-2016-Datasets/td-p/231"><em>Zheng</em> 2016</a> paper (Bold cell types are those that were progressively substituted in setA):</p>
<ul>
<li>setA 100 cell randomly selected for each cell type:
<ul>
<li><ol start="2" style="list-style-type: upper-alpha">
<li>B-cells (25K reads/cell), (M) Monocytes (100K reads/cell), (S) Stem cells (24.7K reads/cell), (NK) Natural Killer cells (29K reads/cell), (N) Naive T-cells (19K reads/cell)</li>
</ol></li>
</ul>
</li>
<li>setB 100 cell randomly selected for each cell type:
<ul>
<li><ol start="2" style="list-style-type: upper-alpha">
<li>B-cells, (M) Monocytes, (H) <strong>T-helper cells</strong> (21K reads/cell), (NK) Natural Killer, (N) Naive T-cell</li>
</ol></li>
</ul>
</li>
<li>setC 100 cell randomly selected for each cell type:
<ul>
<li><ol start="3" style="list-style-type: upper-alpha">
<li>
<strong>Cytotoxic T-cells</strong> (28.6K reads/cell), (M) Monocytes, <strong>T-helper cells</strong>, (NK) Natural Killer, (N) Naive T-cell</li>
</ol></li>
</ul>
</li>
<li>setD 100 cell randomly selected for each cell type:
<ul>
<li><ol start="3" style="list-style-type: upper-alpha">
<li>
<strong>Cytotoxic T-cells</strong>, (NC) <strong>Naive cytotoxic T-cells</strong> (20K reads/cell), (H) <strong>T-helper cells</strong>, (NK) Natural Killer, (N) Naive T-cell</li>
</ol></li>
</ul>
</li>
</ul>
<p>Moving from SetA to setD we added progressively cells coming from T-cell populations, making the cell-type partitioning more challenging because of the similarities between T-cell sub-populations.</p>
<p>We used PCA (Figure below, panels A-D) to visualize the dissimilarity between cells populations. PCA measures the variance between the elements of the dataset and the most important differences in variance are estimated by the PC1.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget https://130.192.119.59/public/section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"cd section4.1_examples"</span>)

<span class="co">#visualizing the complexity of the datasets using PCA</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(docker4seq)
<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"bmsnkn_5x100cells.txt"</span>,<span class="dt">threshold=</span><span class="dv">1000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"bmsnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp), <span class="st">"_"</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">2</span>])
tmp.n1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(tmp.n, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(tmp.n)),<span class="st">"_"</span>,tmp.n, <span class="dt">sep=</span><span class="st">""</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp) &lt;-<span class="st"> </span>tmp.n1
logtmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(tmp <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(logtmp, <span class="st">"log10_bmsnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"log10_bmsnkn_5x100cells_1000.txt"</span>, <span class="dt">type=</span><span class="st">"FPKM"</span>,
    <span class="dt">legend.position=</span><span class="st">"topleft"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
    <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>,
    <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())

<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"bmHnkn_5x100cells.txt"</span>,<span class="dt">threshold=</span><span class="dv">1000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"bmHnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp), <span class="st">"_"</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">2</span>])
tmp.n1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(tmp.n, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(tmp.n)),<span class="st">"_"</span>,tmp.n, <span class="dt">sep=</span><span class="st">""</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp) &lt;-<span class="st"> </span>tmp.n1
logtmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(tmp <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(logtmp, <span class="st">"log10_bmHnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"log10_bmHnkn_5x100cells_1000.txt"</span>, <span class="dt">type=</span><span class="st">"FPKM"</span>,
    <span class="dt">legend.position=</span><span class="st">"topleft"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
    <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>,
    <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())

<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"CmHnkn_5x100cells.txt"</span>,<span class="dt">threshold=</span><span class="dv">1000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"CmHnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp), <span class="st">"_"</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">2</span>])
tmp.n1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(tmp.n, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(tmp.n)),<span class="st">"_"</span>,tmp.n, <span class="dt">sep=</span><span class="st">""</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp) &lt;-<span class="st"> </span>tmp.n1
logtmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(tmp <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(logtmp, <span class="st">"log10_CmHnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"log10_CmHnkn_5x100cells_1000.txt"</span>, <span class="dt">type=</span><span class="st">"FPKM"</span>,
    <span class="dt">legend.position=</span><span class="st">"topleft"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
    <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>,
    <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())

<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"CNCHnkn_5x100cells.txt"</span>,<span class="dt">threshold=</span><span class="dv">1000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)
tmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"CNCHnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp), <span class="st">"_"</span>)
tmp.n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(tmp.n, <span class="cf">function</span>(x)x[<span class="dv">2</span>])
tmp.n1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(tmp.n, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(tmp.n)),<span class="st">"_"</span>,tmp.n, <span class="dt">sep=</span><span class="st">""</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(tmp) &lt;-<span class="st"> </span>tmp.n1
logtmp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>(tmp <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(logtmp, <span class="st">"log10_CNCHnkn_5x100cells_1000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names =</span> <span class="ot">NA</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/docker4seq/topics/pca">pca</a></span>(<span class="dt">experiment.table=</span><span class="st">"log10_CNCHnkn_5x100cells_1000.txt"</span>, <span class="dt">type=</span><span class="st">"FPKM"</span>,
    <span class="dt">legend.position=</span><span class="st">"topleft"</span>, <span class="dt">covariatesInNames=</span><span class="ot">TRUE</span>, <span class="dt">samplesName=</span><span class="ot">FALSE</span>,
    <span class="dt">principal.components=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="dt">pdf =</span> <span class="ot">TRUE</span>,
    <span class="dt">output.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())</code></pre></div>
<p></p>
<div class="figure" style="text-align: center">
<img src="../inst/img/griph1.png" alt="PCA is getting progressively unable to discriminate between the different cell subpopulations as the set of cells are getting functionally more similar to each other: A) PCA of setA, B) PCA of setB, C) PCA of setC, D) PCA of setD" width="80%"><p class="caption">
PCA is getting progressively unable to discriminate between the different cell subpopulations as the set of cells are getting functionally more similar to each other: A) PCA of setA, B) PCA of setB, C) PCA of setC, D) PCA of setD
</p>
</div>
<p>PC1 shows that, as the differences between cell populations become smaller, moving from setA to setD, the aggregation in homogeneous groups of cells is compromised (Figure above, panels A-D).</p>
<p>To observe the effect of the reduced dissimilarity between populations on the stability of the number of clusters, we use the rCASC <strong>clusterNgriph</strong> function on the above mentioned 4 datasets using 160 permutations/each, and randomly removing in each permutation 10% of the cells. Each analysis took approximately 60 mins on a SeqBox hardware.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#downloading datasets</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"section4.1_examples"</span>)

<span class="co">#setA</span>
<span class="kw"><a href="../reference/clusterNgriph.html">clusterNgriph</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), 
              <span class="st">"bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, 
              <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)

<span class="co">#setB</span>
<span class="kw"><a href="../reference/clusterNgriph.html">clusterNgriph</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), 
              <span class="st">"bmHnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nPerm=</span><span class="dv">160</span>,
              <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)

<span class="co">#setC</span>
<span class="kw"><a href="../reference/clusterNgriph.html">clusterNgriph</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), 
              <span class="st">"CmHnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nPerm=</span><span class="dv">160</span>,
              <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)

<span class="co">#setD</span>
<span class="kw"><a href="../reference/clusterNgriph.html">clusterNgriph</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), 
              <span class="st">"CNCHnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nPerm=</span><span class="dv">160</span>,
              <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)</code></pre></div>
<p></p>
<p>It is notable that as the differences between the cells populations is narrowing (i.e in setA cells types are quite different in overall functional activity, as in setD four out of five cell types are T-cells sub-populations) the fluctuations in the detected number of clusters increase. In setA, where PCA is able to discriminate between the five cell populations (Figure below, panel A), out of 160 bootstraps only 1 gave a number of clusters different from the effective number of cell sub-populations (Figure below, panel A). In all the other three subsets perturbations result in a higher number of events in which number of clusters different from 5 (Figure below, panels B-D).</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/griph2.png" alt="Clusters number is dependent by the cell type similarity: A) number of clusters detectable by griph in setA, B) number of clusters detectable by griph in setB, C) number of clusters detectable by griph in setC, D) number of clusters detectable by griph in setD" width="80%"><p class="caption">
Clusters number is dependent by the cell type similarity: A) number of clusters detectable by griph in setA, B) number of clusters detectable by griph in setB, C) number of clusters detectable by griph in setC, D) number of clusters detectable by griph in setD
</p>
</div>
<p>As highlighted in <strong>Section 3.2.1</strong>, the difficulties in detecting a stable number of clusters, upon dataset perturbation, is due to the limited number of detected genes. To further support this hypothesis we run a comparison between the most expressed genes in the cell types used in the above example.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#downloading datasets</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"section4.1_examples"</span>)

<span class="co">#loading the datasets used to generate PCA</span>
b &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"cd19_b_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
mono &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"cd14_mono_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
stem &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"cd34_stem_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
nk &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"cd56_nk_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
naiveT &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"naiveT_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
cyto &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"cytoT_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
naiveCyto &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"naiveCytoT_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
helper &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"cd4_h_100cell.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)

<span class="co">#calculating the gene-level expression and ranking the genes from the most expressed to the least expressed</span>
b.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(b,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)
mono.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(mono,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)
stem.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(stem,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)
nk.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(nk,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)
naiveT.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(naiveT,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)
cyto.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(cyto,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)
naiveCyto.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(naiveCyto,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)
helper.s &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sort">sort</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(helper,<span class="dv">1</span>,sum), <span class="dt">decreasing=</span>T)

<span class="co">#function that measure the identity between lists of increasing lengths</span>
overlap &lt;-<span class="st"> </span><span class="cf">function</span>(x,y){
  overlap.v &lt;-<span class="st"> </span><span class="ot">NULL</span>
  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(x)){
     overlap.v[i] &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sets">intersect</a></span>(x[<span class="dv">1</span><span class="op">:</span>i], y[<span class="dv">1</span><span class="op">:</span>i]))
  }
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(overlap.v)
}

<span class="co">#calculating the level of identity between lists of increasing lengths all comparisons are run with respect to naive T-cell.</span>
naiveCyto.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveCyto.s))
b.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(b.s))
mono.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(mono.s))
stem.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(stem.s))
nk.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(nk.s))
naiveCyto.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveCyto.s))
helper.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(helper.s))
cyto.naiveT &lt;-<span class="st"> </span><span class="kw">overlap</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(naiveT.s), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(cyto.s))

<span class="co">#plotting the above data</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"black"</span>, <span class="dt">lty=</span><span class="dv">2</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), naiveCyto.naiveT[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"black"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), cyto.naiveT[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"blue"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), helper.naiveT[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"green"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), mono.naiveT[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"red"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), b.naiveT[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"brown"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), stem.naiveT[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"orange"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/points">points</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">1</span>, <span class="dv">500</span>), nk.naiveT[<span class="dv">1</span><span class="op">:</span><span class="dv">500</span>], <span class="dt">type=</span><span class="st">"l"</span>, <span class="dt">col=</span><span class="st">"violet"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/legend">legend</a></span>(<span class="st">"topleft"</span>, <span class="dt">legend=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Naive T-cytotoxic"</span>, <span class="st">"T-cytotoxic"</span>, <span class="st">"T-helper"</span>, <span class="st">"Monocytes"</span>, <span class="st">"B-cells"</span>, <span class="st">"Stem cells"</span>, <span class="st">"NK"</span>),
       <span class="dt">pch=</span><span class="dv">15</span>, <span class="dt">col=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"black"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"red"</span>, <span class="st">"brown"</span>, <span class="st">"orange"</span>, <span class="st">"violet"</span>))</code></pre></div>
<p></p>
<p>Figure below shows the number of identical genes found in common between naive T-cells and the other sub-populations in setA and setD, using lists of increasing size and ordered by expression level. The plot shows that naive T-cytotoxic, T-cytoxic and T-helper, from setD, share with naive T-cells, within the top 500 most expressed genes, more genes with respect to the other cell types present in setA. Thus, the lack of cell-type specific genes between 4 out 5 cell types in SetD dataset negatively affect the stability dataset partitioning, upon bootstraps.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/griph4.jpeg" alt="Identity between naive T-cells and the other cell types in set A and D in gene lists of increasing length. Identity between two data sets is shown by the dashed line." width="60%"><p class="caption">
Identity between naive T-cells and the other cell types in set A and D in gene lists of increasing length. Identity between two data sets is shown by the dashed line.
</p>
</div>
<p>The results described in this section indicate that <strong>clusterNgriph</strong> is a valuable instrument to define a range of numbers of clusters to be further investigated with supervised clustering approaches.</p>
</div>
</div>
</div>
<div id="section-5-k-mean-clustering-detecting-cell-sub-populations-by-mean-of-kernel-based-similarity-learning-simlr-" class="section level2">
<h2 class="hasAnchor">
<a href="#section-5-k-mean-clustering-detecting-cell-sub-populations-by-mean-of-kernel-based-similarity-learning-simlr-" class="anchor"></a><strong>Section 5</strong> K-mean clustering: detecting cell sub-populations by mean of kernel based similarity learning (<a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>SIMLR</em></a>).</h2>
<p>The number of clustering and dimension reduction methods for single cell progressively increased over the last few years. Last year <a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>Wang and coworkers</em></a> published SIMLR, a framework which learns a similarity measure from single-cell RNA-seq data in order to perform dimensions reduction. We decided to select this method as core clustering tool in rCASC, because outperformed eight methods published before 2017 [<a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>Wang and coworkers</em></a>].</p>
<p>Specifically, we use SIMLR as clustering method recording the effects of data perturbation, i.e. removal of random subset of cells, on the clustering structure. Although, we think SIMLR provides important advantages with respect to other clustering methods, rCASC framework can embed also other data reduction tools. At the present time, tSne is also implemented within the rCASC data permutation framework.</p>
<p>One of the peculiarity of rCASC is the user tunable bootstrap procedure. rCASC represents bootstrap results via a cell stability score (Figure below). In brief, a set of cells to be organized in clusters (Figure below, panel A) is analysed with SIMLR, applying a users defined k number of clusters (Figure below, panel B). A user defined % of cells is removed from the original data set and these cells are clustered again (Figure below, panel C). The clusters obtained in each bootstrap step are compared with the clusters generated on the full dataset using Jaccard index (Figure below, panels D-E). If the Jaccard index is greater of a user defined threshold, e.g. 0.8, the cluster is called confirmed in the bootstrap step (Figure below, panel F). Then to each cell, belonging to the confirmed cluster, <em>cell stability score</em> value is increased of 1 unit (Figure below, panel G). At the end of the bootstrap procedure, cells are labeled with different symbols describing their <em>cell stability score</em> in a specific cluster (Figure below, panel H).</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/simlr_2.jpeg" alt="Cell stability score" width="60%"><p class="caption">
Cell stability score
</p>
</div>
<p>SIMLR is embedded in <strong>simlrBootstrap</strong> function within the rCASC bootstrap framework.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui18.png" alt="GUI: SIMLR clustering panel" width="70%"><p class="caption">
GUI: SIMLR clustering panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Counts table)</em>, the path of the file, with file name and extension included</li>
<li>
<em>nPerm</em>, number of permutations to be executed</li>
<li>
<em>permAtTime</em>, number of permutations computed in parallel</li>
<li>
<em>percent</em>, percentage of randomly selected cells removed in each permutation</li>
<li>
<em>range1</em>, beginning of the range of clusters to be investigated</li>
<li>
<em>range2</em>, end of the range of clusters to be investigated</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>logTen</em>, 1 if the count matrix is already in log10, 0 otherwise</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input, default is 111</li>
<li>
<em>sp</em>, minimun number of percentage of cells that has to be in common in a cluster, between two permutations, default 0.8</li>
<li>
<em>clusterPermErr</em>, probability error in depicting the number of clusters in each permutation, default = 0.05</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"section4.1_examples"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="co">#annotating data setA</span>
<span class="co">#annotating data setA</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget ftp://ftp.ensembl.org/pub/release-94/gtf/homo_sapiens/Homo_sapiens.GRCh38.94.gtf.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"gzip -d Homo_sapiens.GRCh38.94.gtf.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv Homo_sapiens.GRCh38.94.gtf genome.gtf"</span>)

<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"bmsnkn_5x100cells.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"genome.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">TRUE</span>, <span class="dt">ribo.proteins=</span><span class="ot">TRUE</span>,<span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">riboStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">riboEnd.percentage=</span><span class="dv">100</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, <span class="dt">mitoEnd.percentage=</span><span class="dv">100</span>)
<span class="co">#running SIMLR analysis using the range of clusters suggested by clusterNgriph in session 4.2</span>
<span class="kw"><a href="../reference/simlrBootstrap.html">simlrBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
               <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
               <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">range1=</span><span class="dv">4</span>, <span class="dt">range2=</span><span class="dv">6</span>, 
               <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)

<span class="co">#annotating data setB</span>
<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"bmHnkn_5x100cells.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"genome.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">TRUE</span>, <span class="dt">ribo.proteins=</span><span class="ot">TRUE</span>,<span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">riboStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">riboEnd.percentage=</span><span class="dv">100</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, <span class="dt">mitoEnd.percentage=</span><span class="dv">100</span>)
<span class="co">#running SIMLR analysis using the range of clusters suggested by clusterNgriph in session 4.2</span>
<span class="kw"><a href="../reference/simlrBootstrap.html">simlrBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
               <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_bmHnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
               <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">range1=</span><span class="dv">4</span>, <span class="dt">range2=</span><span class="dv">6</span>, 
               <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)</code></pre></div>
<p></p>
<p>The output of <strong>simlrBootstrap</strong> function is described in Figure . The output is localized in the <em>Results</em> folder (Figure below, Folder 1), which is created in the folder where the analysis started. This folder contains the count matrices used in the analyses. In this example the count matrices are those belonging to setA and B (see Section 4.1). In the <em>Results</em> folder are also present folders labeled with the name of the count matrix used in the analysis (Figure  Folder 2, annotated_bmsnkn_5x100cells, annotated_bmHnkn_5x100cells). In each of these folders there are a set of folders indicated with a number that refers to the analysed range of number of clusters (Figure below, Folder 3). In this specific example the range of clusters goes from 4 to 6. In each <em>NameOfCountMatrix</em> folder there is a file called <strong>_Stability_Violin_Plot.pdf</strong> (Figure below, panel A) which represents the distribution of the cells stability scores over the bootstraps in the range of number of clusters investigated. In this example, it is clear that the analysis done with 5 clusters is the one providing the highest stability of cells within each cluster. The other plot that is also available in the folder (Figure below, Folders 4) is <strong>NameOfCountMatrix_vioplot.pdf</strong>, Figure below, panel B, which contains the distribution of the <a href="https://en.wikipedia.org/wiki/Silhouette_(clustering)"><em>Silhuette</em></a> values for each cluster over the bootstraps. <a href="https://en.wikipedia.org/wiki/Silhouette_(clustering)"><em>Silhuette</em></a> value is a measure of computation cluster stability, and evaluates how similar an object is to its own cluster (cohesion) compared to other clusters (separation). Clearly the information provided by Silhuette plot is much less informative for the definition of the optimal clustering number with respect to the information provided by the cells stability score (Figure below blue arrow). In each clustering folder there is a pdf named <em>NameOfCountMatrix_Stability_Plot.pdf</em>, which contains two plots (Figure below, panels C-D) generated with the clustering program. The plot in Figure below, panel C provides a 2D view of the clustering results. In this plot each cell is labeled with a symbol indicating its cell stability score. In this specific example of the plot generated with 5 clusters shows that all cells remain in a cluster between 75 to 100% of the bootstraps (+ symbol). The plot, Figure below, panel D, shows the genes detectable in each cell in function of the total number of reads/cell. In this plot cells are colored with the same color of their belonging cluster. This plot is useful to observe if the clustering is biased by the number of genes called in each cluster. In this specific example, only the blue cluster is charaterized by a number of detected genes, which is larger of those detectable in the other clusters (Figure below, panel D). This is expected, since blue cluster is made of Monocytes, which have been sequenced to 100K reads/cells, as all other cells in setA were sequenced between 19 to 29K reads/cell, see <strong>Section 4.2</strong>.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/simlr_3.jpeg" alt="simlrBootstrap output" width="80%"><p class="caption">
simlrBootstrap output
</p>
</div>
<p>In Figure below are shown the effects, on cell stability, in SIMLR analysis done with 4 and 6 clusters for SetA (<em>annotated_bmsnkn_5x100cells.txt</em>) counts matrix. Using 4 as number of clusters, Figure below, panel A, some of the cells in each cluster show a reduced stability (50-75%, triangle) and arrows highlights cells characterized by a cell stability score between 25 to 50%. The circles in Figure below, panel B show the clusters where the full cluster has a cell stability between 50 to 75%. This observation indicates that the stability score is a useful measure to identify the optimal number of partitions to be used, i.e. the highest cell stability score was observed when five clusters were selected, corresponding to the number of cell types combined in SetA, see <strong>Section 4.1</strong>.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/simlr_4.jpeg" alt="Clustering output including cell stability score" width="80%"><p class="caption">
Clustering output including cell stability score
</p>
</div>

<div id="section-5-2-visualizing-the-cell-clusters-relocation-during-bootstraps-" class="section level3">
<h3 class="hasAnchor">
<a href="#section-5-2-visualizing-the-cell-clusters-relocation-during-bootstraps-" class="anchor"></a><strong>Section 5.2</strong> Visualizing the cell clusters relocation during bootstraps.</h3>
<p>The function <strong>permutationMovie</strong> allows the generation of a video showing the relocation of cells at each bootstrap. In this example we use the results shown in Figure above, panel B, where circles show the clusters where all cells have a cell stability between 50 to 75%.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui19.png" alt="GUI: Permutation effect  video panel" width="70%"><p class="caption">
GUI: Permutation effect video panel
</p>
</div>
<ul>
<li>
<p><em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):</p>
<ul>
<li>
<em>scratch.folder</em>, path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, path of the file, with file name and extension included</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>framePP</em>, number of frames for each permutation</li>
<li>
<em>permutationNumber,</em> number of random permutations, must be less or the same value of the total permutations dome in simlrBootstrap</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/test_permutationvideo.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"test_permutationvideo.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"test_permutationvideo"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"rCASC"</span>)
<span class="kw"><a href="../reference/bootstrapsVideo.html">bootstrapsVideo</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>, 
                <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                <span class="dt">nCluster=</span><span class="dv">6</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">framePP=</span><span class="dv">200</span>, <span class="dt">permutationNumber=</span><span class="dv">80</span>)</code></pre></div>
<p></p>
<p>The video was generated in 8 minutes on SeqBox hardware. The video is saved in the cluster folder used for the analysis and it has the name <strong>outputname.mp4</strong>. The video generated with the above script is accessible <a href="https://youtu.be/ZBP9W-WU-WQ"><em>here</em></a>.</p>
<p>In Figure below, panel A is shown a screenshot of the output of <strong>bootstrapsVideo</strong>. The output of this function provides extra information with respect to the standard output of the <strong>simlrBootstrap</strong> and <strong>tsneBootstrap</strong>, since the video provides an overview of the area (colored circles) in which the cells of a specific cluster are localized as consequence of the bootstraps. This visualization provides a better description of the maximum size of the clusters and simplify the identification of clusters that are more near to each other, because of clusters structure, Figure below, panel A. In this specific example, is notable that, even if the yellow cluster has a high cell stability score, Figure below, panel B, its size is much greater of those observable for all the other clusters, because few cells with lower cell stability score (50-75%, triangle) are located very near to the blue cluster, arrow in Figure below, panel B.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/bootstrapsVideo_2.jpeg" alt="Comparing the output of bootstrapsVideo with respect to the one produced by simlrBootstrap. A) Output of bootstrapsVideo. B) simlrBootstrap output. " width="100%"><p class="caption">
Comparing the output of bootstrapsVideo with respect to the one produced by simlrBootstrap. A) Output of bootstrapsVideo. B) simlrBootstrap output.
</p>
</div>

</div>
</div>
<div id="section-6-hierarchical-clustering-seurat-graph-based-clustering-" class="section level2">
<h2 class="hasAnchor">
<a href="#section-6-hierarchical-clustering-seurat-graph-based-clustering-" class="anchor"></a><strong>Section 6</strong> Hierarchical clustering: <a href="https://satijalab.org/seurat/pbmc3k_tutorial.html"><em>Seurat</em></a> graph-based clustering.</h2>
<p>Recently <a href="https://f1000research.com/articles/7-1297/v1"><em>Freytag et al.</em></a> and <a href="https://f1000research.com/articles/7-1141/v2"><em>Do et al.</em></a> described, in their independent comparison papers, that <a href="https://satijalab.org/seurat/"><em>Seurat</em></a> delivered the overall best performance in cells clustering. Thus, we decided to include in our workflow also this clustering tool. Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metagene’ that combines information across a correlated gene set. The first step of the analysis is the identification of the how many PCs have to be to included.</p>
<p><strong>seuratPCAEval</strong> function allows the exploration of PCs to determine relevant sources of heterogeneity and to define the range of PCs to be used:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui20.png" alt="GUI: Estimating PCA range for Seurat panel" width="70%"><p class="caption">
GUI: Estimating PCA range for Seurat panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, the path of the file, with file name and extension included</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>logTen</em>, 1 if the count matrix is already in log10, 0 otherwise</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input, default is 111</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"section4.1_examples"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="co">#defining the PC threshold for clustering.</span>
<span class="kw"><a href="../reference/seuratPCAEval.html">seuratPCAEval</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
              <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
              <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">logTen =</span> <span class="dv">0</span>, <span class="dt">seed =</span> <span class="dv">111</span>)</code></pre></div>
<p></p>
<div class="figure" style="text-align: center">
<img src="../inst/img/seurat01.png" alt="Seurat clustering, A) Seurat dataset PC component. B) Clustering results" width="50%"><p class="caption">
Seurat clustering, A) Seurat dataset PC component. B) Clustering results
</p>
</div>
<p>The function generate a <strong>Result</strong> folder in which is present a folder with the same name of the dataset under analysis, in this specific case “bmsnkn_5x100cells”. In the latter folder is present a PCA plot (PCE_bowPlot.pdf, Figure above, panel A), which allow the identification of the PCs to use, defining a cutoff where there is a clear elbow in the graph. In this example, it looks like the elbow would fall around PC 6.</p>
<p><strong>seuratBootstrap</strong> function executes the seurat clustering (data reduction method PCA) and estimates cell stability score:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui21.png" alt="GUI: Seurat clustering panel" width="70%"><p class="caption">
GUI: Seurat clustering panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, a character string indicating the path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, a character string indicating the path of the file, with file name and extension included</li>
<li>
<em>nPerm</em>, number of permutations to be executed</li>
<li>
<em>permAtTime</em>, number of permutations computed in parallel</li>
<li>
<em>percent</em>, percentage of randomly selected cells removed in each permutation</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>logTen</em>, 1 if the count matrix is already in log10, 0 otherwise</li>
<li>
<em>pcaDimensions</em>, 0 for automatic selection of PC elbow.</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/seuratBootstrap.html">seuratBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">pcaDimensions=</span><span class="dv">6</span>, <span class="dt">seed=</span><span class="dv">111</span>)</code></pre></div>
<p></p>
<p>The function generate a <strong>Result</strong> folder in which is present a folder with the same name of the dataset under analysis, in this specific case “bmsnkn_5x100cells.txt”. In the latter folder is present a folder with the number of cluster detected by Seurat, in this specific case <strong>5</strong>. In <strong>5</strong> folder there is the file with extension <strong>_Stability_Plot.pdf</strong>, which is the clustering picture with cells labeled with their stability score). The file with the extension <strong>_clusterP.txt</strong>,containing for each permutation the location of each cell in the clusters. The file with the extension <strong>_killedCell.txt</strong>, where are saved the cells removed at each permutation. The file with the extension <strong>_score.txt</strong>, containing the scores assigned at each permutation to each cell. The file with extension <strong>_scoreSum.txt</strong> containing the Cell stability score for each cell. The file with extension <strong>_clustering.output.txt</strong>, summarizing all information required to generate the <strong>_Stability_Plot.pdf</strong> plot.</p>
<div id="section-6-1-comparing-simlr-tsne-and-seurat-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#section-6-1-comparing-simlr-tsne-and-seurat-clustering" class="anchor"></a><strong>Section 6.1</strong> Comparing SIMLR, tSne and Seurat clustering</h3>
<p>In Wang [<a href="https://www.ncbi.nlm.nih.gov/pubmed/28263960"><em>2017</em></a>] article it is demonstrated that SIMLR outperforms other data reduction methods. <a href="https://f1000research.com/articles/7-1297/v1"><em>Freytag et al.</em></a> and <a href="https://f1000research.com/articles/7-1141/v2"><em>Do et al.</em></a> described that <a href="https://satijalab.org/seurat/"><em>Seurat</em></a> delivers the overall best performance in cells clustering. Here, we run a comparison between SIMLR, tSne and Seurat clustering (data reduction by PCA) within the rCASC bootstrap framework to observe how cells stability score is affected by the three clustering methods.</p>
<p>tSne is embedded in <strong>tsneBootstrap</strong> function within the rCASC bootstrap framework.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui22.png" alt="GUI: tSne clustering panel" width="70%"><p class="caption">
GUI: tSne clustering panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, the path of the file, with file name and extension included</li>
<li>
<em>nPerm</em>, number of permutations to be executed</li>
<li>
<em>permAtTime</em>, number of permutations computed in parallel</li>
<li>
<em>percent</em>, percentage of randomly selected cells removed in each permutation</li>
<li>
<em>range1</em>, beginning of the range of clusters to be investigated</li>
<li>
<em>range2</em>, end of the range of clusters to be investigated</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>logTen</em>, 1 if the count matrix is already in log10, 0 otherwise</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input, default is 111</li>
<li>
<em>sp</em>, minimun number of percentage of cells that has to be in common in a cluster, between two permutations, default 0.8</li>
<li>
<em>clusterPermErr</em>, probability error in depicting the number of clusters in each permutation, default = 0.05</li>
<li>
<em>perplexity</em>, number of close neighbors for each point. This parameter is specific for tSne. Default value is 10.</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"section4.1_examples"</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget ftp://ftp.ensembl.org/pub/release-94/gtf/homo_sapiens/Homo_sapiens.GRCh38.94.gtf.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"gzip -d Homo_sapiens.GRCh38.94.gtf.gz"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mv Homo_sapiens.GRCh38.94.gtf genome.gtf"</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)
<span class="co">#annotating data setA</span>

<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"bmsnkn_5x100cells.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"genome.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">TRUE</span>, <span class="dt">ribo.proteins=</span><span class="ot">TRUE</span>,<span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">riboStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">riboEnd.percentage=</span><span class="dv">100</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, <span class="dt">mitoEnd.percentage=</span><span class="dv">100</span>)

<span class="co">#running tSne analysis using the range of clusters suggested by clusterNgriph in session 4.2</span>
<span class="kw"><a href="../reference/tsneBootstrap.html">tsneBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
              <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
              <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">range1=</span><span class="dv">4</span>, <span class="dt">range2=</span><span class="dv">6</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,
              <span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>, <span class="dt">sp=</span><span class="fl">0.8</span>, <span class="dt">perplexity=</span><span class="dv">10</span>)
<span class="co"># required time 159 mins</span>

<span class="co">#running SIMLR analysis using the range of clusters suggested by clusterNgriph in session 4.2</span>
<span class="kw"><a href="../reference/simlrBootstrap.html">simlrBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
               <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
               <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">range1=</span><span class="dv">4</span>, <span class="dt">range2=</span><span class="dv">6</span>, 
               <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)
<span class="co"># required time 168 mins</span>

<span class="co">#running Seurat clustering: defining the PC threshold for clustering.</span>
<span class="kw"><a href="../reference/seuratPCAEval.html">seuratPCAEval</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
              <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
              <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">logTen =</span> <span class="dv">0</span>, <span class="dt">seed =</span> <span class="dv">111</span>)
<span class="co">#running Seurat clustering with bootstrap: threshold was detected at PC 6, see Section 6.</span>
<span class="kw"><a href="../reference/seuratBootstrap.html">seuratBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">pcaDimensions=</span><span class="dv">6</span>, <span class="dt">seed=</span><span class="dv">111</span>)
<span class="co"># required time 194 mins</span></code></pre></div>
<p></p>
<p>We run the analysis on the SetA described in Section 4.1. Both SIMLR and tSne are able to detect that five clusters provides the more stable organization in terms of cells stability score (Figure below, panels A-B). However, SIMLR clustering produces highly stable clusters (Figure below, panel C), as instead tSne clustering produces much less stable clusters (Figure below, panel D). From the point of view of the computation time the two clustering methods are relatively similar. The above mentioned analysis took 159 mins for tSne and 168 mins for SIMLR. Seurat clustering detected 5 clusters, Figure B, as SIMLR and tSne. The cell stability score for SIMLR (Figure below, panel C) and Seurat (Figure below, panel B) are superimposable and computation time is very similar to the other two methods.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/simlr_5.jpeg" alt="Comparing SIMLR and tSne within rCASC bootstrap framework. A,C) SetA clustered with SIMLR. B,D) SetA clustered with tSne." width="100%"><p class="caption">
Comparing SIMLR and tSne within rCASC bootstrap framework. A,C) SetA clustered with SIMLR. B,D) SetA clustered with tSne.
</p>
</div>

</div>
</div>
<div id="section-7-detecting-the-genes-playing-the-major-role-in-clusters-formation" class="section level2">
<h2 class="hasAnchor">
<a href="#section-7-detecting-the-genes-playing-the-major-role-in-clusters-formation" class="anchor"></a><strong>Section 7</strong> Detecting the genes playing the major role in clusters formation</h2>
<p>Nowaday there are a lot of methods for the identification of differentally expressed genes between two populations of single-cell experiments [<a href="https://www.ncbi.nlm.nih.gov/pubmed/29481549"><em>Soneson et al.</em></a>]. However, the number of tools applicable to single-cells and allowing something similar to an ANOVA are very few. An ANOVA-like approach is described as applicable to single-cells in <a href="https://www.ncbi.nlm.nih.gov/pubmed/19910308"><em>edgeR Bioconductor package</em></a>. The edgeR ANOVA-like requires a comparison with respect to a reference set of cells. SIMLR and Seurat also provide ranking score for the genes, which are affecting mostly the clusters organization. SIMLR and Seurat gene ranking do not require a reference cluster for the analysis, thus providing wider applications with respect to ANOVA-like. The three methods are part of the rCASC framework:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui23a.png" alt="GUI: Feature selection panel" width="30%"><p class="caption">
GUI: Feature selection panel
</p>
</div>
<div id="section-7-1-a-statistical-approach-to-select-genes-affecting-clusters-formation-edger-anova-like" class="section level3">
<h3 class="hasAnchor">
<a href="#section-7-1-a-statistical-approach-to-select-genes-affecting-clusters-formation-edger-anova-like" class="anchor"></a><strong>Section 7.1</strong> A statistical approach to select genes affecting clusters formation: EdgeR ANOVA-like</h3>
<p>The function <strong>anovaLike</strong> executes edgeR ANOVA-like with the settings required for single-cells analysis, for more information please refer to the edgeR vignette.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui23.png" alt="GUI: ANOVA-like feature selection panel" width="70%"><p class="caption">
GUI: ANOVA-like feature selection panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>file (GUI Matrix count)</em>, a character string indicating the counts table file with the path of the file.</li>
<li>
<em>sep</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>cluster.file</em>, a character string indicating the _clustering.output.txt file of interest, generated by bootstrapSimlar or bootStrapTsne. IMPORTANT this file must be located in the same folder where counts.table is placed</li>
<li>
<em>ref.cluster</em>, a number indicating the cluster to be used a reference for anova-like comparison with the other clusters.</li>
<li>
<em>logFC.threshold</em>, minimal logFC present in at least one of the comparisons with respect to reference covariate</li>
<li>
<em>FDR.threshold</em>, minimal FDR present in at least one of the comparisons with respect to reference covariate</li>
<li>
<em>logCPM.threshold</em>, minimal average abundance</li>
<li>
<em>plot</em>, boolean, TRUE a plot of differentially expressed genes is generated</li>
</ul>
</li>
</ul>
<p>The input file is a counts table, which is modified by <strong>anovaLike</strong> adding the cluster number to cell name with an underscore. It is mandatory that no other underscores are located in the table. Then, counts table is reorganized to locate at the beginning of the table the set of cells belonging to the cluster used as reference group. The differentially expressed genes statistics for all genes are saved in the file with prefix <strong>DE_</strong>, followed by the counts table name. The filtered differentially expressed genes statistics, including only genes detected as differentially expressed in at least one of the clusters is saved with prefix <strong>filtered_DE_</strong> followed by the counts table name. <strong>logFC_filtered_DE_</strong> refers to the file containing only logFC extracted from <strong>filtered_DE_</strong> file. The count table is also saved reordered on the basis of cluster positions and has the extension <strong>_reordered.txt</strong></p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/annotated_setPace_10000_noC5_clustering.output.txt"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/annotated_setPace_10000_noC5.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"annotated_setPace_10000_noC5.txt.zip"</span>)

<span class="kw"><a href="../reference/anovaLike.html">anovaLike</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="dt">counts.table=</span><span class="st">"annotated_setPace_10000_noC5.txt"</span>,
        <span class="dt">file.type=</span><span class="st">"txt"</span>, <span class="dt">cluster.file=</span><span class="st">"annotated_setPace_10000_noC5_clustering.output.txt"</span>, <span class="dt">ref.cluster=</span><span class="dv">3</span>,
        <span class="dt">logFC.threshold=</span><span class="dv">1</span>, <span class="dt">FDR.threshold=</span><span class="fl">0.05</span>, <span class="dt">logCPM.threshold=</span><span class="dv">4</span>, <span class="dt">plot=</span><span class="ot">TRUE</span>)

<span class="co">#the output of interest is logFC_filtered_DE_annotated_setPace_10000_noC5_reordered.txt</span></code></pre></div>
<p></p>
<p>The function <strong>clustersFeatures</strong> selects the genes, which are more up-regulated in a specific cluster using <strong>anovaLike</strong> outputs:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui24.png" alt="GUI: ANOVA like prioritization features panel" width="70%"><p class="caption">
GUI: ANOVA like prioritization features panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>fileLogFC</em>, a character string indicating the absolute path to logFC_filtered_DE_ file generated by anovaLike function</li>
<li>
<em>fileCounts</em>, a character string indicating the absolute path to reordered counts table file generated by anovaLike function</li>
<li>
<em>delta</em>, the minimal distance between the max value of FC for a gene in a cluster of interest and the nearest other max FC in any of the other clusters. This value define the minimal distance with respect to the same gene in an other cluster to identify it as a cluster specific gene.</li>
<li>
<em>sep</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
</ul>
</li>
</ul>
<p>This function focus only on upregulated genes because the ANOVA-like procedure compares all clusters with respect to a reference cluster. Thus, up-regulated genes are genes specific of the cluster under analysis as those down-regulated are characteristics of the reference cluster. The delta parameter describes the minimal distance existing, for a specific gene average logFC in a specific cluster, with respect to all the other clusters under analysis. The output of the function are four tab delimited files:</p>
<ul>
<li><p>the file with prefix <strong>onlyUP_</strong>, followed by the counts table name, i.e. the count table only containing the selected genes,</p></li>
<li>the file with prefix<strong>onlyUP_</strong>, followed by <strong>logFC_filtered_DE_</strong>, the table containing logFC only for the selected genes,</li>
<li><p>the file <strong>onlyUP_clusters_specific_genes.txt</strong>, which contains the list of specific genes associated with the corresponding cluster.</p></li>
<li><p>the file <strong>onlyUP_clusters_specific_genesSYMBOLs.txt</strong>, which contains the list of specific genes SYMBOLS from onlyUP_clusters_specific_genes.txt. This file is used by function <strong>hfc</strong> to generate features specific heatmap.</p></li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/clusters.features.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"clusters.features.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"clusters.features"</span>)

<span class="kw"><a href="../reference/clustersFeatures.html">clustersFeatures</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),
         <span class="dt">logFC.table=</span><span class="st">"logFC_filtered_DE_annotated_setPace_10000_noC5_reordered.txt"</span>,
         <span class="dt">counts.table=</span><span class="st">"annotated_setPace_10000_noC5_reordered.txt"</span>, <span class="dt">delta=</span><span class="fl">0.5</span>)</code></pre></div>
<p></p>
<div id="section-7-1-1-visualizing-cluster-specific-genes-by-heatmap-" class="section level4">
<h4 class="hasAnchor">
<a href="#section-7-1-1-visualizing-cluster-specific-genes-by-heatmap-" class="anchor"></a><strong>Section 7.1.1</strong> Visualizing cluster specific genes by heatmap.</h4>
<p>The gene-features extracted by <strong>clustersFeatures</strong> cannot be used again in <strong>simlrBootstrap</strong>, if the overall number of genes is smaller of the cell number, for more information see SIMLR publication. This limitation does not apply to tsneBootstrap.</p>
<p>The function <strong>hfc</strong> allows to create a heatmap for the results generated from <strong>simrBootstrap/tsneBootstrap</strong> and the output of <strong>clustersFeatures</strong>.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui25.png" alt="GUI: Heatmap for prioritized genes panel" width="70%"><p class="caption">
GUI: Heatmap for prioritized genes panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, the path of the file, with file name and extension included</li>
<li>
<em>nCluster</em>, n of the clusters on which to run the analysis. The function expect the presence of Result folder generated by <strong>simrBootstrap/tsneBootstrap</strong>
</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>lfn</em>, name of the list of genes WITHOUT extension, if <strong>clustersFeatures</strong> function was used on anovaLike output the name of the file is “onlyUP_clusters_specific_geneSYMBOLs”</li>
<li>
<em>geneNameControl</em>, 0 if the matrix has gene name without ENSEMBL geneID. 1 if the gene names is formatted like this : ENSMUSG00000000001:Gnai3.</li>
<li>
<em>status</em>, 0 if is raw count, 1 otherwise</li>
<li>
<em>b1</em>, the lower range of signal in the heatmap,for negative value write “/-5”. To ask the function to do automatic value assignment set 0</li>
<li>
<em>b2</em>, the upper range of signal in the heatmap,for negative value write “/-2”. To ask the function to do automatic value assignment set 0</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/hfc.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"clusters.features.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"clusters.features"</span>)

<span class="kw"><a href="../reference/hfc.html">hfc</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>,
       <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"annotated_setPace_10000_noC5.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
       <span class="dt">nCluster=</span><span class="dv">5</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">lfn=</span><span class="st">"onlyUP_clusters_specific_geneSYMBOLs"</span>, 
       <span class="dt">geneNameControl=</span><span class="dv">1</span>, <span class="dt">status=</span><span class="dv">0</span>, <span class="dt">b1=</span><span class="st">"/-1"</span>, <span class="dt">b2=</span><span class="st">"1"</span>)</code></pre></div>
<p></p>
<p>The output is made of four pdf:</p>
<ul>
<li><p><em>lfn-parameter</em>_hfc_log10.pdf counts in log10 format, e.g. onlyUP_clusters_specific_geneSYMBOLs_hfc_log10.pdf</p></li>
<li><p><em>lfn-parameter</em>_hfc_zscore.pdf Z score</p></li>
<li><p><em>lfn-parameter</em>_hfc_CSS.pdf cell cluster stability score</p></li>
<li><p><em>lfn-parameter</em>_hfc_all_genes.pdf</p></li>
</ul>
<div class="figure" style="text-align: center">
<img src="../inst/img/hfc1.jpeg" alt="Output of hfc. A) Path of location of hfc function results. The output is located in cluster folder used for the analysis, B) Log10 counts heatmap. On the top of the figure clusters are represented as bars of different colors, C) Z-score heatmap, D) Cell Stability Score for each cell in each cluster, for more information see Section 5, E) SIMLR clustering output for each gene within those used to generate the heatmap. Cells are colored on the basis of the amount of CPMs for the gene under analysis." width="100%"><p class="caption">
Output of hfc. A) Path of location of hfc function results. The output is located in cluster folder used for the analysis, B) Log10 counts heatmap. On the top of the figure clusters are represented as bars of different colors, C) Z-score heatmap, D) Cell Stability Score for each cell in each cluster, for more information see Section 5, E) SIMLR clustering output for each gene within those used to generate the heatmap. Cells are colored on the basis of the amount of CPMs for the gene under analysis.
</p>
</div>
</div>
</div>
<div id="section-7-2-a-machine-learning-approach-simlr-genes-prioritization" class="section level3">
<h3 class="hasAnchor">
<a href="#section-7-2-a-machine-learning-approach-simlr-genes-prioritization" class="anchor"></a><strong>Section 7.2</strong> A machine learning approach: SIMLR genes prioritization</h3>
<p>SIMLR gene prioritization approach relies on the output similarity and does not depend on the downstream dimension reduction or clustering. The advantage of this step is that one can identify highly fluctuating genes that are consistent with the learned similarity in the multiple-kernel space.</p>
<p>The function <strong>genesPrioritization</strong> executes SIMLR gene prioritization within the rCASC bootstrap framework.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui26.png" alt="GUI: SIMLR genes prioritization panel" width="70%"><p class="caption">
GUI: SIMLR genes prioritization panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, the path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, the path of the file, with file name and extension included</li>
<li>
<em>nPerm</em>, number of permutations to be executed</li>
<li>
<em>permAtTime</em>, number of permutations that can be computes in parallel</li>
<li>
<em>percent</em>, percentage of randomly selected cells removed in each permutation</li>
<li>
<em>nCluster</em>, the number of clusters, where to run prioritization</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>logTen</em>, 1 if the count matrix is already in log10, 0 otherwise</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input</li>
<li>
<em>sp</em>, minimun number of percentage of cells that has to be in common between two permutation to be the same cluster, default 0.8.</li>
<li>
<em>clusterPermErr</em>, error that can be done by each permutation in cluster number depicting, default=0.05</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/annotated_setPace_10000_noC5.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"annotated_setPace_10000_noC5.txt.zip"</span>)
<span class="kw"><a href="../reference/genesPrioritization.html">genesPrioritization</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
                    <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"annotated_setPace_10000_noC5.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                    <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">nCluster=</span><span class="dv">5</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, 
                    <span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>, <span class="dt">sp=</span><span class="fl">0.8</span>, <span class="dt">clusterPermErr=</span><span class="fl">0.05</span>)</code></pre></div>
<p></p>
<p>The output of the function are a pdf named <strong>geneRank.pdf</strong>, containing the overall distribution of -log10 expression mean with respect to the Delta confidence of a gene being important for clustering, and a tab delimited file with the extension <strong>_pvalList.txt</strong> which contains, for each bootstrap, the gene importance score for clustering.</p>
<p>The function <strong>genesSelection</strong> allows to run a selection of the most interesting genes for clustering on the basis of two parameters:</p>
<ul>
<li>maxDeltaConfidence:
<ul>
<li>max value for Delta confidence for genes feature selection. Default is 0.01</li>
</ul>
</li>
<li>minLogMean:
<ul>
<li>min value for Log mean for genes feature selection. Default is 0.05</li>
</ul>
</li>
</ul>
<p>The above values can be set in the <strong>genesSelection</strong> function on the basis of the output of <strong>geneRank.pdf</strong>.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui27.png" alt="GUI: SIMLR genes selection panel" width="70%"><p class="caption">
GUI: SIMLR genes selection panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, the path of the counts file, with file name and extension included</li>
<li>
<em>nCluster</em>, the number of clusters, where prioritization was run. In the working folder should be present the Result folder generated by <strong>genesPrioritization</strong> function.</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input, default 111</li>
<li>
<em>sp</em>, minimun number of percentage of cells that has to be in common in a cluster, between two permutations, default 0.8</li>
<li>
<em>clusterPermErr</em>, probability error in depicting the number of clusters in each permutation, default = 0.05</li>
<li>
<em>maxDeltaConfidence</em>, max value for Delta confidence for genes feature selection</li>
<li>
<em>minLogMean</em>, min value for Log mean for genes feature selection</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/genesSelection.html">genesSelection</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_setPace_10000_noC5.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nCluster=</span><span class="dv">5</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">seed=</span><span class="dv">111</span>, <span class="dt">sp=</span><span class="fl">0.8</span>, <span class="dt">clusterPermErr=</span><span class="fl">0.05</span>, <span class="dt">maxDeltaConfidence=</span><span class="fl">0.01</span>, <span class="dt">minLogMean=</span><span class="dv">20</span>)

<span class="co">#mv annotated_setPace_10000_noC5_clusters_specific_geneSYMBOLs.txt in the main folder  </span>

<span class="kw"><a href="../reference/hfc.html">hfc</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>,
       <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"annotated_setPace_10000_noC5.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
       <span class="dt">nCluster=</span><span class="dv">5</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, 
       <span class="dt">lfn=</span><span class="st">"annotated_setPace_10000_noC5_clusters_specific_geneSYMBOLs"</span>, 
       <span class="dt">geneNameControl=</span><span class="dv">1</span>, <span class="dt">status=</span><span class="dv">0</span>, <span class="dt">b1=</span><span class="st">"/-1"</span>, <span class="dt">b2=</span><span class="st">"1"</span>)</code></pre></div>
<p></p>
<div class="figure" style="text-align: center">
<img src="../inst/img/geneselection.jpeg" alt="genePrioritization and genesSelection outputs. A) Prioritization plot, from this plot the maxDeltaConfidence (x axis) and the minLogMean (y axis), required by genesSelection can be extrapolated, i.e. in this case respectively 0.01 and 20. B) log10 counts heatmap from the output of genesSelection: 577 genes. C) Z-score heatmap. D) Cell stability score. E) simlrBootstrap output. F) Single gene CPM expression in SIMLR clusters. The expression is available for each of the genes from genesSelection output" width="100%"><p class="caption">
genePrioritization and genesSelection outputs. A) Prioritization plot, from this plot the maxDeltaConfidence (x axis) and the minLogMean (y axis), required by genesSelection can be extrapolated, i.e. in this case respectively 0.01 and 20. B) log10 counts heatmap from the output of genesSelection: 577 genes. C) Z-score heatmap. D) Cell stability score. E) simlrBootstrap output. F) Single gene CPM expression in SIMLR clusters. The expression is available for each of the genes from genesSelection output
</p>
</div>
<p>The output of <strong>genesSelection</strong> is a file with extension **_clusters_specific_geneSYMBOLs.txt<strong>, which can be used as input for </strong>hfc** function, see <strong>Section 6.1.1</strong>. The list of prioritized genes cannot not be associated to a specific cluster as in the case of the anovaLike, see <strong>Section 6.1</strong>. Thus, more information will be provided on the cluster specificity from the heatmap visualization, see <strong>hfc</strong> function.</p>
</div>
<div id="section-7-3-seurat-genes-prioritization" class="section level3">
<h3 class="hasAnchor">
<a href="#section-7-3-seurat-genes-prioritization" class="anchor"></a><strong>Section 7.3</strong> Seurat genes prioritization</h3>
<p>Seurat gene prioritization is performed by <strong>seuratPrior</strong> function.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/gui28.png" alt="GUI: Seurat genes prioritization panel" width="70%"><p class="caption">
GUI: Seurat genes prioritization panel
</p>
</div>
<ul>
<li>
<em>Parameters</em> (only those without default; for the full list of parameters please refer to the function help) (Figure above):
<ul>
<li>
<em>scratch.folder</em>, a character string indicating the path of the scratch folder</li>
<li>
<em>file (GUI Matrix count)</em>, a character string indicating the path of the file, with file name and extension included</li>
<li>
<em>separator</em>, separator used in count file, e.g. ‘\t’, ‘,’</li>
<li>
<em>logTen</em>, 1 if the count matrix is already in log10, 0 otherwise</li>
<li>
<em>seed</em>, important value to reproduce the same results with same input</li>
<li>
<em>PCADim</em>, dimensions of PCA for seurat clustering</li>
<li>
<em>geneNumber</em>, numbers of specific genes for each clusters</li>
<li>
<em>nCluster</em>, number of cluster analysis.</li>
</ul>
</li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"section4.1_examples.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"section4.1_examples"</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rCASC)

<span class="co">#running Seurat PC distribution</span>
<span class="kw"><a href="../reference/seuratPCAEval.html">seuratPCAEval</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
              <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
              <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">logTen =</span> <span class="dv">0</span>, <span class="dt">seed =</span> <span class="dv">111</span>)

<span class="co">#running Seurat clustering with bootstrap</span>
<span class="kw"><a href="../reference/seuratBootstrap.html">seuratBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">pcaDimensions=</span><span class="dv">6</span>, <span class="dt">seed=</span><span class="dv">111</span>)

<span class="co">#running Seurat clustering with bootstrap: elbow was detected at PC 6, see Section 6.</span>
<span class="kw"><a href="../reference/seuratPrior.html">seuratPrior</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>, 
            <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"bmsnkn_5x100cells.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, 
            <span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>, <span class="dt">PCADim=</span><span class="dv">6</span>, <span class="dt">geneNumber=</span><span class="dv">20</span>, <span class="dt">nCluster=</span><span class="dv">5</span>)</code></pre></div>
<p></p>
<p>The output is a file with the prefix <strong>Markers_</strong>:</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/seurat02.png" alt="Seurat gene prioritization" width="70%"><p class="caption">
Seurat gene prioritization
</p>
</div>

</div>
</div>
<div id="section-8-an-example-of-rcasc-analysis-geogse106264-" class="section level2">
<h2 class="hasAnchor">
<a href="#section-8-an-example-of-rcasc-analysis-geogse106264-" class="anchor"></a><strong>Section 8</strong> An example of rCASC analysis: <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE106264"><em>GEO:GSE106264</em></a>.</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/29326266">Pace et al.</a> explored the role of histone methyltransferase Suv39h1 in murine CD8+ T cells activated after Listeria monocytogenes infection. They showed that Suv39h1 controls the expression of a set of stem cell-related memory genes and its silencing increases long-term memory reprogramming capacity. The single-cell sequencing data from the <a href="https://www.ncbi.nlm.nih.gov/pubmed/29326266"><em>Pace’s Science paper</em></a>:</p>
<ul>
<li><p>naive CD8+ T lymphocytes, here named <strong>N</strong></p></li>
<li><p>CD8+ T cells activated after Listeria monocytogenes infection, here named <strong>NA</strong></p></li>
<li><p>naive Suv39h1-defective CD8+ T lymphocytes, here named <strong>Nd</strong></p></li>
<li><p>Suv39h1-defective CD8+ T cells activated after Listeria monocytogenes infection, here named <strong>NdA</strong></p></li>
</ul>
<div id="section-8-1-selecting-subsets-of-cells-with-the-highest-number-of-called-genes-" class="section level3">
<h3 class="hasAnchor">
<a href="#section-8-1-selecting-subsets-of-cells-with-the-highest-number-of-called-genes-" class="anchor"></a><strong>Section 8.1</strong> Selecting subsets of cells with the highest number of called genes.</h3>
<p>On the basis of our observation in <strong>Section 3.2.1</strong>, we selected, for each of the above mentioned groups, the cells with higher number of total reads/cell. Specifically we analysed a total of 600 cells, combining respectively 50 cells from <strong>N</strong> and <strong>Nd</strong> and 250 cells for <strong>NA</strong> and <strong>NdA</strong>. The rational of the different number of cells for the naive and the activated is due to the expectation that activated cells will be organize in mutiple clusters and naive are expected to be homogeneous. Cell labels in the counts table were modified adding the groups <em>N</em>, <em>Nd</em>, <em>NA</em> and <em>NdA</em>.</p>
<p>Using this dataset we tried to address the following questions:</p>
<ol style="list-style-type: decimal">
<li><p>Does Suv39h1 silence gene expression programs in naive CD8+ T cells?</p></li>
<li><p>Does Suv39h1 silence gene expression programs in activated CD8+ T cells?</p></li>
<li><p>Does Suv39h1 silencing affect the differentiation of new subsets of activated CD8+ T lymphocytes?</p></li>
</ol>
<p>With the script below we have performed the following steps:</p>
<ul>
<li><p>annotating the counts table with <strong>scannobyGt</strong> function,</p></li>
<li><p>removing the ribosomal and mitochondrial genes with <strong>scannobyGt</strong> function,</p></li>
<li><p>selecting the most expressed 10K genes for the analysis with <strong>topx</strong> function,</p></li>
<li><p>detecting the range of interesting number of cluster to be used for data partitioning (6-9) with <strong>clusterNgriph</strong></p></li>
<li><p>identifying 6 as the optimal number of clusters and evaluated the cell stability in each cluster with <strong>simlrBootstrap</strong></p></li>
<li><p>generating the video of the bootstrap analysis with <strong>bootstrapsVideo</strong></p></li>
</ul>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/setPace.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"setPace.txt.zip"</span>)

<span class="co">#annotating genes and removing ribosomal and mitochondrial proteins</span>
<span class="kw"><a href="../reference/scannobyGtf.html">scannobyGtf</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="st">"setPace.txt"</span>,<span class="dt">sep=</span><span class="st">"/"</span>),
                    <span class="dt">gtf.name=</span><span class="st">"genome_mm10.gtf"</span>, <span class="dt">biotype=</span><span class="st">"protein_coding"</span>, 
                    <span class="dt">mt=</span><span class="ot">FALSE</span>, <span class="dt">ribo.proteins=</span><span class="ot">FALSE</span>,<span class="dt">umiXgene=</span><span class="dv">3</span>, <span class="dt">riboStart.percentage=</span><span class="dv">0</span>, 
                    <span class="dt">riboEnd.percentage=</span><span class="dv">100</span>, <span class="dt">mitoStart.percentage=</span><span class="dv">0</span>, <span class="dt">mitoEnd.percentage=</span><span class="dv">100</span>)

<span class="co">#selecting 10K top expressed genes</span>
<span class="kw"><a href="../reference/topx.html">topx</a></span>(<span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),<span class="dt">file.name=</span><span class="st">"annotated_setPace.txt"</span>,<span class="dt">threshold=</span><span class="dv">10000</span>, <span class="dt">logged=</span><span class="ot">FALSE</span>)

<span class="co">#defining the range of number of clusters to be investigated</span>
<span class="kw"><a href="../reference/clusterNgriph.html">clusterNgriph</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),
              <span class="st">"annotated_setPace_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nPerm=</span><span class="dv">160</span>,
              <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)
<span class="co">#38 mins on SeqBox hardware</span>

<span class="co">#running the SIMLR clustering in the range 6-9 detected by clusterNgriph</span>
<span class="kw"><a href="../reference/simlrBootstrap.html">simlrBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
               <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_setPace_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
               <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">range1=</span><span class="dv">6</span>, <span class="dt">range2=</span><span class="dv">9</span>, 
               <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)

<span class="co"># visualizing cells instability</span>
<span class="kw"><a href="../reference/bootstrapsVideo.html">bootstrapsVideo</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">scratch.folder=</span><span class="st">"/data/scratch"</span>, 
                <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_setPace_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
                <span class="dt">nCluster=</span><span class="dv">6</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">framePP=</span><span class="dv">200</span>, <span class="dt">permutationNumber=</span><span class="dv">80</span>)</code></pre></div>
<p></p>

<div class="figure" style="text-align: center">
<img src="../inst/img/pace_1.jpeg" alt="Analysis of a subset of cells from GSE106264 dataset. A) Clusters detected by clusterNgriph. B) Cell stability score detected by simlrBootstrap. C) Clusterization with 6 clusters with simlrBootstrap, in parentesis is given the overall stability score of each cluster. The components of each cluster, in terms of cells experiment groups, are also indicated for the most stable clusters. D) Permutation 55 extracted from the video generate with bootstrapsVideo." width="100%"><p class="caption">
Analysis of a subset of cells from GSE106264 dataset. A) Clusters detected by clusterNgriph. B) Cell stability score detected by simlrBootstrap. C) Clusterization with 6 clusters with simlrBootstrap, in parentesis is given the overall stability score of each cluster. The components of each cluster, in terms of cells experiment groups, are also indicated for the most stable clusters. D) Permutation 55 extracted from the video generate with bootstrapsVideo.
</p>
</div>
<p>In Figure above are summarized the results of the analysis executed on the Pace’s dataset. Data perturbations, Figure above, panel A, allows data organization between 6 to 9 clusters, where 7 clusters is the most represented group. Cell stability score, from the SIMLR analysis executed on the above range of clusters, is shown in Figure above, panel B. Six clusters show a slighly higher stability with respect to the others. The overall stability of 6 clusters is however sub-optimal, since it is spread between 0 and 0.9 cell stability score. In Figure above, panel C it is shown the clusters structure generated with SIMLR on 6 clusters. Clusters 1, 3 and 4 show a quite good stability, Figure above, panel C. Cluster 3 is made of 44 <strong>N</strong> (88%) and 48 <strong>Nd</strong> (96%), suggesting that naive CD8+ T lymphocytes are not affected by the silencing of Suv39h1 gene. Cluster 1 contains 16 <strong>NA</strong> (6.4%) and 14 <strong>NdA</strong> (5.6%). Cluster 2 is made of 44.8% of <strong>NA</strong> and 39.6% of <strong>NdA</strong> cells. Clusters 1 and 2 group together, interdependently from Suv39h1 gene silencing. Cluster 6 is made of 35% <strong>NA</strong> and 13.6% of <strong>NdA</strong>. In cluster 6 the amount of <strong>NA</strong> and <strong>NdA</strong> is unbalance, suggesting that the Suv39h1 silencing does not guarantee the efficient differentiation of the cell subpopulation in cluster 6. Cluster 4 only contains <strong>NdA</strong> (33%), indicating that at least a subpopulation of activated Suv39h1-silenced cells has a specific transcription profile that differentiate them from the wild type activated cells. Cluster 5 is made of 6 <strong>N</strong> cells, 2 <strong>Nd</strong> cells, 34 <strong>NA</strong> cells, 21 <strong>NdA</strong> cells. Despite the presence of a limited amount of naive cells, which might be explained as partially activated, cluster 5 is made mainly of activated cells, i.e. 13.6% NA and 8.4% NdA of total cells. The cluster structure (Figure above, panel D) and cell cluster stability scores (Figure above, panel C) might suggest that cluster 5 is made of a precursor subset.</p>
<p><strong>NB</strong> % always refers to the total number of cells used in this example (50 N, 50 Nd, 250 NA, 250 NdA); <a href="http://130.192.119.59/public/annotated_setPace_10000_clustering.output.txt"><em>annotated_setPace_10000_clustering.output.txt</em></a>, contains all information required to reproduce clusters shown in Figure above, panel C. In Figure above, panel D it is shown permutation 55 of the output of <strong>bootstrapsVideo</strong>, which can be seen <a href="https://youtu.be/w5Jj8QxKXRU"><em>here</em></a>. This permutation is representative of the majority of the observed permutations. From <strong>bootstrapsVideo</strong> analysis is clear that cluster 5 is strongly affecting the cell score stability of clusters 6 and 2 ad for less extent that of cluster 3.</p>
<p>This analysis can provide some preliminary answers to the questions raised above:</p>
<ol style="list-style-type: decimal">
<li>
<p>Does Suv39h1 silence gene expression programs in naive CD8+ T cells?</p>
<ol style="list-style-type: lower-alpha">
<li>No, naive cells cluster together independently from Suv39h1 silencing.</li>
</ol>
</li>
<li>
<p>Does Suv39h1 silence gene expression programs in activated CD8+ T cells?</p>
<ol style="list-style-type: lower-alpha">
<li>Yes, cluster 4 is only made of an activated Suv39h1 KO sub-population. Furthermore, although quite unstable, cluster 6 sub-population is less represented in activated Suv39h1 silenced cells, suggesting a reduction in efficiency in the differentiation of this sub-population upon Suv39h1 silencing.</li>
</ol>
</li>
<li>
<p>Does Suv39h1 silencing affect the differentiation of new subsets of activated CD8+ T lymphocytes?</p>
<ol style="list-style-type: lower-alpha">
<li>Yes, Cluster 4 represents a unique Suv39h1 subset of activated CD8+ T cells.</li>
</ol>
</li>
</ol>
</div>
<div id="section-8-2-improving-clustering-results" class="section level3">
<h3 class="hasAnchor">
<a href="#section-8-2-improving-clustering-results" class="anchor"></a><strong>Section 8.2</strong> Improving clustering results</h3>
<p>On the basis of Figure D and the output of <a href="https://youtu.be/w5Jj8QxKXRU"><em>bootstrapsVideo</em></a> clearly cells in cluster 5 represent a strong interference for an efficient clustering. Thus, we investigated if, removing cluster 5 cells (Figure above, panel C), cell stability score for cluster 6 and 2 could be improved.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/annotated_setPace_10000.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"annotated_setPace_10000.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/annotated_setPace_10000_clustering.output.txt"</span>)

<span class="co">#removing cluster 5 cells from annotated_setPace_10000.txt</span>

pace &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"annotated_setPace_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)
clusters.info &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="st">"annotated_setPace_10000_clustering.output.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">header=</span>T, <span class="dt">row.names=</span><span class="dv">1</span>)

discard &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(clusters.info)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(clusters.info<span class="op">$</span>Belonging_Cluster<span class="op">==</span><span class="dv">5</span>)]

pace &lt;-<span class="st"> </span>pace[,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(pace)<span class="op">%in%</span>discard)]
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/write.table">write.table</a></span>(pace, <span class="st">"annotated_setPace_10000_noC5.txt"</span>, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">col.names=</span><span class="ot">NA</span>)

<span class="co">#defining the range of number of clusters to be investigated</span>
<span class="kw"><a href="../reference/clusterNgriph.html">clusterNgriph</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,<span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),
              <span class="st">"annotated_setPace_10000_noC5.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), <span class="dt">nPerm=</span><span class="dv">160</span>,
              <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)


<span class="co">#running the SIMLR clustering in the range 5-9 detected by clusterNgriph</span>
<span class="kw"><a href="../reference/simlrBootstrap.html">simlrBootstrap</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>,<span class="dt">scratch.folder=</span><span class="st">"/data/scratch/"</span>,
               <span class="dt">file=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="st">"annotated_setPace_10000.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>), 
               <span class="dt">nPerm=</span><span class="dv">160</span>, <span class="dt">permAtTime=</span><span class="dv">8</span>, <span class="dt">percent=</span><span class="dv">10</span>, <span class="dt">range1=</span><span class="dv">5</span>, <span class="dt">range2=</span><span class="dv">9</span>, 
               <span class="dt">separator=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,<span class="dt">logTen=</span><span class="dv">0</span>, <span class="dt">seed=</span><span class="dv">111</span>)
<span class="co">#execution time 276 mins in SeqBox</span></code></pre></div>
<p></p>
<p>The results of the analysis done removing cells associated to cluster 5 are shown in Figure . The range of clusters suggested by <strong>clusterNgriph</strong> are between 5 and 9, Figure below, panel A. The best cells stability score is observable for 5 clusters, Figure below, panel B. <strong>simlrBootstrap</strong> results show a large improvment of the overall stablity of the clusters Figure below, panel C with respect to the previous analysis, Figure above, panel C. It is also notable that the samples organization with in clusters agrees with preliminary answers in Section 5.1.1 .</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/pace_2.jpeg" alt="Analysis of Pace dataset. A) Clusters detected by clusterNgriph. B) Cell stability score detected by simlrBootstrap. C) Clusterization with 5 clusters with simlrBootstrap, in parentesis is given the overall stability score of each cluster. The components of each cluster, in terms of cells experiment groups, are also indicated for the most stable clusters." width="100%"><p class="caption">
Analysis of Pace dataset. A) Clusters detected by clusterNgriph. B) Cell stability score detected by simlrBootstrap. C) Clusterization with 5 clusters with simlrBootstrap, in parentesis is given the overall stability score of each cluster. The components of each cluster, in terms of cells experiment groups, are also indicated for the most stable clusters.
</p>
</div>
</div>
<div id="section-8-3-detecting-the-genes-playing-the-major-role-in-clusters-generation-edger-anova-like-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#section-8-3-detecting-the-genes-playing-the-major-role-in-clusters-generation-edger-anova-like-analysis" class="anchor"></a><strong>Section 8.3</strong> Detecting the genes playing the major role in clusters generation: EdgeR ANOVA-like analysis</h3>
<p>We detected a total of 2742 differentially expressed genes using the naive cluster 3 as reference in an ANOVA-like analysis, see <strong>Section 6.1</strong>.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/annotated_setPace_10000_noC5_clustering.output.txt"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/annotated_setPace_10000_noC5.txt.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"annotated_setPace_10000_noC5.txt.zip"</span>)

<span class="kw"><a href="../reference/anovaLike.html">anovaLike</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(), <span class="dt">counts.table=</span><span class="st">"annotated_setPace_10000_noC5.txt"</span>,
        <span class="dt">file.type=</span><span class="st">"txt"</span>, <span class="dt">cluster.file=</span><span class="st">"annotated_setPace_10000_noC5_clustering.output.txt"</span>, <span class="dt">ref.cluster=</span><span class="dv">3</span>,
        <span class="dt">logFC.threshold=</span><span class="dv">1</span>, <span class="dt">FDR.threshold=</span><span class="fl">0.05</span>, <span class="dt">logCPM.threshold=</span><span class="dv">4</span>, <span class="dt">plot=</span><span class="ot">TRUE</span>)

<span class="co">#the output of interest is logFC_filtered_DE_annotated_setPace_10000_noC5_reordered.txt</span></code></pre></div>
<p></p>
<p>The following analysis focus only on up-regulated genes, i.e. those gene specifically more expressed in a cluster with respect to the naive cluster used as reference in the ANOVA-like.</p>
<p></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"wget http://130.192.119.59/public/clusters.features.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/unzip">unzip</a></span>(<span class="st">"clusters.features.zip"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">setwd</a></span>(<span class="st">"clusters.features"</span>)

<span class="kw"><a href="../reference/clustersFeatures.html">clustersFeatures</a></span>(<span class="dt">group=</span><span class="st">"docker"</span>, <span class="dt">data.folder=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>(),
         <span class="dt">logFC.table=</span><span class="st">"logFC_filtered_DE_annotated_setPace_10000_noC5_reordered.txt"</span>,
         <span class="dt">counts.table=</span><span class="st">"annotated_setPace_10000_noC5_reordered.txt"</span>, <span class="dt">delta=</span><span class="fl">0.5</span>)</code></pre></div>
<p></p>
<p>We analysed the results coming out of the <strong>clustersFeatures</strong> selection focusing on the two clusters that are different in cell ratio between wt and Suv39h1-defective T-cells: cluster 1, showing a 2.4:1 ratio between wt and Suv39h1-defective T-cells, and 2, only detected in Suv39h1-defective T-cells, Figure above, panel C. 31 genes were detected as specific for cluster 1. <a href="http://amp.pharm.mssm.edu/Enrichr/"><em>EnrichR</em></a> analysis indicated the overepresentation of STAT3 in ENCODE and ChEA Consensus TFs from ChIP-X datasets and TNFRSF13B, CCL5, CCL4 and CXCR6 in Cytokine-cytokine receptor interaction dataset. On the basis of the above observations and on a PUBMED search (<a href="https://www.ncbi.nlm.nih.gov/pubmed/16982880"><em>Ccl5</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/23830147"><em>STAT3</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/27883012"><em>Ccl9</em></a> and <a href="https://www.ncbi.nlm.nih.gov/pubmed/28855310"><em>Cxcr6</em></a>), we suggest that cluster 1 is made of early activated precursors having some traits of memory cells. This cluster is also characterized by an unbalance ratio between activated wt T-cells and Suv39h1-defective T-cells.</p>
<p>The cluster 2 has some traits of memory cells, on the basis of EnrichR analysis of 100 genes. Enrichment analysis shows enrichment for STAT3 transcription specific paths and GO Biological process enrichment for <strong>response to cytokine</strong>. Furthermore, a PUBMED search showed the presence in this cluster of memory specific genes such as <a href="https://www.ncbi.nlm.nih.gov/pubmed/28973925"><em>TCF7</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/28973925"><em>FOXO1</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/24335726"><em>Bach2</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/18000950"><em>Ccr7</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/24335726"><em>Id3</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/25567129"><em>Il7r</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/16449532"><em>Myc</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/28137869"><em>Pdk1</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/22118528"><em>Socs3</em></a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed/16177108"><em>Tnfsf8</em></a>,</p>
<p>The analysis of cluster 2 was extended using Ingenuity Pathways Analysis (IPA). This sub-population of cells, only detectable within Suv39h1-defective T-cells, is enriched for immuno-related functions: Development of hematopoietic progenitor cells, Adhesion of lymphocytes, Quantity of T-lymphocytes, systemic autoimmune syndrome. A subset of cluster 2 genes can be also aggregated in a network characterized by various transcription factors (Figure below). Interestingly, in this network are present Id3, Bach2 IL7R and Myc. Combining the information that Bach2 is associated to the generation (<a href="https://www-ncbi-nlm-nih-gov.bibliopass.unito.it/pubmed/27158840"><em>Roychoudhuri et al. 2016</em></a>) of long-lived memory cells and Id3 to their maintainance (<a href="https://www-ncbi-nlm-nih-gov.bibliopass.unito.it/pubmed/22057289"><em>Yang et al. 2011</em></a>) together with the knowledge that IL7R/Myc expression is found to be associated to a longer persistence of mouse memory T-cell (<a href="https://www.ncbi.nlm.nih.gov/pubmed/?term=29326266"><em>Pace et al. 2018</em></a>) it coud be suggested that cells belonging to this cluster are precursors of long-lived memory cells.</p>
<div class="figure" style="text-align: center">
<img src="../inst/img/c2_connected.png" alt="IPA C2 connected genes. " width="80%"><p class="caption">
IPA C2 connected genes.
</p>
</div>
<p>The paper <a href="https://www.ncbi.nlm.nih.gov/pubmed/29326266"><em>Pace</em></a> showed that Suv39h1-defective CD8+ T cells show sustained survival and increased long-term memory reprogramming capacity. Our reanalysis with rCASC extends the information provided from the single cell analysis in Pace paper, suggesting the presence of an enriched Suv39h1-defective memory subset, cluster 2. Noteworthy, this example suggests that even few hunders cells are sufficient to discriminate between sub-populations.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#section-1-rcasc-a-single-cell-analysis-workflow-designed-to-provide-data-reproducibility"><strong>Section 1</strong> rCASC: a single cell analysis workflow designed to provide data reproducibility</a></li>
      <li><a href="#section-2-counts-generation"><strong>Section 2</strong> Counts generation</a></li>
      <li><a href="#section-3-counts-matrix-editing"><strong>Section 3</strong> Counts matrix editing</a></li>
      <li><a href="#section-4-estimating-the-number-of-cluster-to-be-used-for-cell-sub-population-discovery-"><strong>Section 4</strong> Estimating the number of cluster to be used for cell sub-population discovery.</a></li>
      <li>
<a href="#section-5-k-mean-clustering-detecting-cell-sub-populations-by-mean-of-kernel-based-similarity-learning-simlr-"><strong>Section 5</strong> K-mean clustering: detecting cell sub-populations by mean of kernel based similarity learning (</a><a href="https://www-ncbi-nlm-nih-gov/pubmed/28263960"><em>SIMLR</em></a>).</li>
      <li>
<a href="#section-6-hierarchical-clustering-seurat-graph-based-clustering-"><strong>Section 6</strong> Hierarchical clustering: </a><a href="https://satijalab-org/seurat/pbmc3k_tutorial-html"><em>Seurat</em></a> graph-based clustering.</li>
      <li><a href="#section-7-detecting-the-genes-playing-the-major-role-in-clusters-formation"><strong>Section 7</strong> Detecting the genes playing the major role in clusters formation</a></li>
      <li>
<a href="#section-8-an-example-of-rcasc-analysis-geogse106264-"><strong>Section 8</strong> An example of rCASC analysis: </a><a href="https://www-ncbi-nlm-nih-gov/geo/query/acc-cgi?acc=GSE106264"><em>GEO:GSE106264</em></a>.</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Raffaele Calogero, Luca Alessandri, Marco Beccuti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
